{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f51a4141",
   "metadata": {},
   "source": [
    "# Enrichment Analysis Notebook\n",
    "\n",
    "Compares the results of SPINDOCTOR gene set summarization vs statistical ontological enrichment.\n",
    "\n",
    "Draft: https://docs.google.com/document/d/1H103ux6Dd1_bPM0un4RwutBLcYJx-0ybil2AwlAvG_Q/edit#"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01aa8c37",
   "metadata": {},
   "source": [
    "## Initial setup\n",
    "\n",
    "Here we take care of imports, defining the data dictionary for the pandas dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c1963df9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pip 23.0.1 from /Users/marcin/Documents/VIMSS/ontology/LLMs/enrichgpt-results/venv/lib/python3.10/site-packages/pip (python 3.10)\n"
     ]
    }
   ],
   "source": [
    "!pip --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "895d256d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Package                    Version\n",
      "-------------------------- ------------\n",
      "adeft                      0.11.2\n",
      "aiohttp                    3.8.5\n",
      "aiosignal                  1.3.1\n",
      "airium                     0.2.5\n",
      "altair                     5.1.1\n",
      "aniso8601                  9.0.1\n",
      "annotated-types            0.5.0\n",
      "antlr4-python3-runtime     4.9.3\n",
      "anyio                      4.0.0\n",
      "appdirs                    1.4.4\n",
      "appnope                    0.1.3\n",
      "argon2-cffi                23.1.0\n",
      "argon2-cffi-bindings       21.2.0\n",
      "arrow                      1.2.3\n",
      "asttokens                  2.4.0\n",
      "async-lru                  2.0.4\n",
      "async-timeout              4.0.3\n",
      "attrs                      23.1.0\n",
      "Babel                      2.12.1\n",
      "backcall                   0.2.0\n",
      "bcp47                      0.0.4\n",
      "beautifulsoup4             4.12.2\n",
      "bioc                       2.1\n",
      "bleach                     6.0.0\n",
      "blinker                    1.6.2\n",
      "boto3                      1.28.45\n",
      "botocore                   1.31.45\n",
      "cachetools                 5.3.1\n",
      "cachier                    2.2.1\n",
      "cattrs                     23.1.2\n",
      "certifi                    2023.7.22\n",
      "cffi                       1.15.1\n",
      "CFGraph                    0.2.1\n",
      "chardet                    5.2.0\n",
      "charset-normalizer         3.2.0\n",
      "class-resolver             0.4.2\n",
      "click                      8.1.7\n",
      "click-default-group        1.2.4\n",
      "click-default-group-wheel  1.2.3\n",
      "colorama                   0.4.6\n",
      "comm                       0.1.4\n",
      "contourpy                  1.1.0\n",
      "curies                     0.6.1\n",
      "cycler                     0.11.0\n",
      "debugpy                    1.7.0\n",
      "decorator                  5.1.1\n",
      "defusedxml                 0.7.1\n",
      "Deprecated                 1.2.14\n",
      "deprecation                2.1.0\n",
      "docopt                     0.6.2\n",
      "dominate                   2.8.0\n",
      "EditorConfig               0.12.3\n",
      "et-xmlfile                 1.1.0\n",
      "eutils                     0.6.0\n",
      "exceptiongroup             1.1.3\n",
      "executing                  1.2.0\n",
      "fastjsonschema             2.18.0\n",
      "fastobo                    0.12.2\n",
      "Flask                      2.1.3\n",
      "Flask-Bootstrap            3.3.7.1\n",
      "flask-restx                1.1.0\n",
      "Flask-WTF                  1.1.1\n",
      "fonttools                  4.42.1\n",
      "fqdn                       1.5.1\n",
      "frozenlist                 1.4.0\n",
      "funowl                     0.2.3\n",
      "ghp-import                 2.1.0\n",
      "gilda                      1.0.0\n",
      "gitdb                      4.0.10\n",
      "GitPython                  3.1.35\n",
      "gpt4                       0.0.1\n",
      "gpt4all                    1.0.10\n",
      "graphviz                   0.20.1\n",
      "greenlet                   2.0.1\n",
      "h11                        0.14.0\n",
      "hbreader                   0.9.1\n",
      "httpcore                   0.18.0\n",
      "httpx                      0.25.0\n",
      "idna                       3.4\n",
      "ijson                      3.2.3\n",
      "importlib-metadata         6.8.0\n",
      "inflect                    7.0.0\n",
      "inflection                 0.5.1\n",
      "iniconfig                  2.0.0\n",
      "intervaltree               3.1.0\n",
      "ipykernel                  6.25.2\n",
      "ipython                    8.15.0\n",
      "isodate                    0.6.1\n",
      "isoduration                20.11.0\n",
      "itsdangerous               2.1.2\n",
      "jedi                       0.19.0\n",
      "Jinja2                     3.1.2\n",
      "jmespath                   1.0.1\n",
      "joblib                     1.3.2\n",
      "jsbeautifier               1.14.9\n",
      "json-flattener             0.1.9\n",
      "json5                      0.9.14\n",
      "jsonasobj                  1.3.1\n",
      "jsonasobj2                 1.0.4\n",
      "jsonlines                  4.0.0\n",
      "jsonpatch                  1.33\n",
      "jsonpath-ng                1.5.3\n",
      "jsonpointer                2.4\n",
      "jsonschema                 4.19.0\n",
      "jsonschema-specifications  2023.7.1\n",
      "jupyter_client             8.3.1\n",
      "jupyter_core               5.3.1\n",
      "jupyter-events             0.7.0\n",
      "jupyter-lsp                2.2.0\n",
      "jupyter_server             2.7.3\n",
      "jupyter_server_terminals   0.4.4\n",
      "jupyterlab                 4.0.5\n",
      "jupyterlab-pygments        0.2.2\n",
      "jupyterlab_server          2.24.0\n",
      "kgcl-rdflib                0.5.0\n",
      "kgcl-schema                0.6.0\n",
      "kiwisolver                 1.4.5\n",
      "lark                       1.1.7\n",
      "linkml                     1.5.7\n",
      "linkml-dataops             0.1.0\n",
      "linkml-owl                 0.3.0\n",
      "linkml-renderer            0.3.0\n",
      "linkml-runtime             1.5.6\n",
      "llm                        0.8.1\n",
      "llm-gpt4all                0.1.1\n",
      "lxml                       4.9.3\n",
      "Markdown                   3.4.4\n",
      "markdown-it-py             3.0.0\n",
      "MarkupSafe                 2.1.3\n",
      "matplotlib                 3.7.2\n",
      "matplotlib-inline          0.1.6\n",
      "mdurl                      0.1.2\n",
      "mergedeep                  1.3.4\n",
      "mistune                    3.0.1\n",
      "mkdocs                     1.5.2\n",
      "mkdocs-material            9.3.1\n",
      "mkdocs-material-extensions 1.1.1\n",
      "mkdocs-mermaid2-plugin     0.6.0\n",
      "more-click                 0.1.2\n",
      "more-itertools             10.1.0\n",
      "multidict                  6.0.4\n",
      "nbclient                   0.8.0\n",
      "nbconvert                  7.8.0\n",
      "nbformat                   5.9.2\n",
      "ndex2                      3.5.1\n",
      "nest-asyncio               1.5.7\n",
      "networkx                   3.1\n",
      "nlpcloud                   1.1.44\n",
      "nltk                       3.8.1\n",
      "notebook                   7.0.3\n",
      "notebook_shim              0.2.3\n",
      "numpy                      1.25.2\n",
      "oaklib                     0.5.19\n",
      "ols-client                 0.1.4\n",
      "ontogpt                    0.3.1\n",
      "ontoportal-client          0.0.4\n",
      "openai                     0.28.0\n",
      "openpyxl                   3.1.2\n",
      "overrides                  7.4.0\n",
      "packaging                  23.1\n",
      "paginate                   0.5.6\n",
      "pandas                     2.1.0\n",
      "pandocfilters              1.5.0\n",
      "pansql                     0.0.1\n",
      "parse                      1.19.1\n",
      "parso                      0.8.3\n",
      "pathspec                   0.11.2\n",
      "pathtools                  0.1.2\n",
      "patsy                      0.5.3\n",
      "pexpect                    4.8.0\n",
      "pickleshare                0.7.5\n",
      "Pillow                     9.5.0\n",
      "pip                        23.0.1\n",
      "platformdirs               3.10.0\n",
      "pluggy                     1.3.0\n",
      "ply                        3.11\n",
      "portalocker                2.7.0\n",
      "prefixcommons              0.1.12\n",
      "prefixmaps                 0.1.5\n",
      "prometheus-client          0.17.1\n",
      "prompt-toolkit             3.0.39\n",
      "pronto                     2.5.5\n",
      "protobuf                   4.24.3\n",
      "psutil                     5.9.5\n",
      "ptyprocess                 0.7.0\n",
      "pure-eval                  0.2.2\n",
      "pyarrow                    13.0.0\n",
      "pycparser                  2.21\n",
      "pydantic                   2.3.0\n",
      "pydantic_core              2.6.3\n",
      "pydeck                     0.8.0\n",
      "Pygments                   2.16.1\n",
      "PyJSG                      0.11.10\n",
      "pymdown-extensions         10.3\n",
      "Pympler                    1.0.1\n",
      "pyparsing                  3.0.9\n",
      "PyShEx                     0.8.1\n",
      "PyShExC                    0.9.1\n",
      "pysolr                     3.9.0\n",
      "pystow                     0.5.0\n",
      "pytest                     7.4.2\n",
      "pytest-logging             2015.11.4\n",
      "python-dateutil            2.8.2\n",
      "python-json-logger         2.0.7\n",
      "python-multipart           0.0.5\n",
      "python-ulid                1.1.0\n",
      "PyTrie                     0.4.0\n",
      "pytz                       2023.3.post1\n",
      "pytz-deprecation-shim      0.1.0.post0\n",
      "PyYAML                     6.0.1\n",
      "pyyaml_env_tag             0.1\n",
      "pyzmq                      25.1.1\n",
      "ratelimit                  2.2.1\n",
      "rdflib                     7.0.0\n",
      "rdflib-jsonld              0.6.1\n",
      "rdflib-shim                1.0.3\n",
      "referencing                0.30.2\n",
      "regex                      2023.8.8\n",
      "requests                   2.31.0\n",
      "requests-cache             1.1.0\n",
      "requests-toolbelt          1.0.0\n",
      "rfc3339-validator          0.1.4\n",
      "rfc3986-validator          0.1.1\n",
      "rfc3987                    1.3.8\n",
      "rich                       13.5.2\n",
      "rpds-py                    0.10.2\n",
      "ruamel.yaml                0.17.32\n",
      "ruamel.yaml.clib           0.2.7\n",
      "s3transfer                 0.6.2\n",
      "scikit-learn               1.3.0\n",
      "scipy                      1.11.2\n",
      "seaborn                    0.12.2\n",
      "semsimian                  0.2.1\n",
      "semsql                     0.3.2\n",
      "Send2Trash                 1.8.2\n",
      "setuptools                 68.2.1\n",
      "ShExJSG                    0.8.2\n",
      "six                        1.16.0\n",
      "smmap                      5.0.0\n",
      "sniffio                    1.3.0\n",
      "sortedcontainers           2.4.0\n",
      "soupsieve                  2.5\n",
      "sparqlslurper              0.5.1\n",
      "SPARQLWrapper              2.0.0\n",
      "SQLAlchemy                 2.0.20\n",
      "SQLAlchemy-Utils           0.38.3\n",
      "sqlite-fts4                1.0.3\n",
      "sqlite-utils               3.35.1\n",
      "sssom                      0.3.41\n",
      "sssom-schema               0.15.0\n",
      "stack-data                 0.6.2\n",
      "statsmodels                0.14.0\n",
      "streamlit                  1.26.0\n",
      "tabulate                   0.9.0\n",
      "tenacity                   8.2.3\n",
      "terminado                  0.17.1\n",
      "threadpoolctl              3.2.0\n",
      "tiktoken                   0.4.0\n",
      "tinycss2                   1.2.1\n",
      "toml                       0.10.2\n",
      "tomli                      2.0.1\n",
      "toolz                      0.12.0\n",
      "tornado                    6.3.3\n",
      "tqdm                       4.66.1\n",
      "traitlets                  5.9.0\n",
      "typing_extensions          4.7.1\n",
      "tzdata                     2023.3\n",
      "tzlocal                    4.3.1\n",
      "Unidecode                  1.3.6\n",
      "uri-template               1.3.0\n",
      "url-normalize              1.4.3\n",
      "urllib3                    1.26.16\n",
      "validators                 0.22.0\n",
      "visitor                    0.1.3\n",
      "watchdog                   3.0.0\n",
      "wcwidth                    0.2.6\n",
      "webcolors                  1.13\n",
      "webencodings               0.5.1\n",
      "websocket-client           1.6.3\n",
      "Werkzeug                   2.1.2\n",
      "wikipedia                  1.4.0\n",
      "Wikipedia-API              0.6.0\n",
      "wrapt                      1.15.0\n",
      "WTForms                    3.0.1\n",
      "yarl                       1.9.2\n",
      "zipp                       3.16.2\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "10e21961-35f1-4c45-bb07-3411ca7cad17",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install --force-reinstall --upgrade more-itertools\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fec960ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/marcin/Documents/VIMSS/ontology/LLMs/enrichgpt-results/venv/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import yaml\n",
    "from yaml import Loader\n",
    "from collections import defaultdict\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import kstest, ttest_ind, wilcoxon\n",
    "import math\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "#import itertools as it\n",
    "#import collections as ct\n",
    "#import more_itertools as mit\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.patches import Rectangle\n",
    "import colorsys\n",
    "import seaborn as sns\n",
    "\n",
    "from oaklib import get_adapter\n",
    "from oaklib.datamodels.vocabulary import IS_A, PART_OF\n",
    "from ontogpt.evaluation.enrichment.eval_enrichment import EvalEnrichment\n",
    "go = get_adapter(\"sqlite:obo:go\")\n",
    "hgnc = get_adapter(\"sqlite:obo:hgnc\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ebcc3593",
   "metadata": {},
   "outputs": [],
   "source": [
    "TURBO = \"gpt-3.5-turbo\"\n",
    "DAVINCI = \"text-davinci-003\"\n",
    "GPT4 = \"gpt-4\"\n",
    "MODELS = [TURBO, DAVINCI, GPT4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fb21ba1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data dictionary\n",
    "MODEL = \"model\"\n",
    "METHOD = \"method\"\n",
    "HAS_TOP_HIT = \"has top term\"\n",
    "IN_TOP_5 = \"in top 5\"\n",
    "IN_TOP_10 = \"in top 10\"\n",
    "RANK = \"rank\"\n",
    "SIZE_OVERLAP = \"size overlap\"\n",
    "SIMILARITY = \"similarity\"\n",
    "NR_SIZE_OVERLAP = \"nr size overlap\"\n",
    "NR_SIMILARITY = \"nr similarity\"\n",
    "GENESET = \"gene_set\"\n",
    "PROMPT_VARIANT = \"prompt_variant\"\n",
    "SOURCE_GENESET = \"source geneset\"\n",
    "GENESET_DESCRIPTION = \"description\"\n",
    "GENESET_SIZE = \"gene_set_size\"\n",
    "TRUNCATION_FACTOR = \"truncation factor\"\n",
    "NUM_TERMS = \"num terms\"\n",
    "NUM_GO_TERMS = \"num GO terms\"\n",
    "UNPARSED = \"unparsed\"\n",
    "NUM_UNPARSED = \"num unparsed\"\n",
    "TERM_IDS = \"term ids\"\n",
    "GO_TERM_IDS = \"go term ids\"\n",
    "GO_TERM_P_VALUES = \"go term p values\"\n",
    "MAX_P_VALUE = \"max p value\"\n",
    "MIN_P_VALUE = \"min p value\"\n",
    "MEAN_P_VALUE = \"mean p value\"\n",
    "PROPORTION_SIGNIFICANT = \"proportion significant\"\n",
    "NOVEL = \"unannotated\"\n",
    "NOVEL_LABELS = \"unannotated labels\"\n",
    "NUM_NOVEL = \"num unannotated\"\n",
    "GENE_RANDOMIZATION_FACTOR = \"gene_randomization_factor\"\n",
    "SUMMARY = \"summary\"\n",
    "TP_num = \"number of true positives across a set of gene GO annotations\"\n",
    "FP_num = \"number of false positives across a set of gene GO annotations\"\n",
    "TP = \"true positive rate across a set of gene GO annotations\"\n",
    "FP = \"false positive rate across a set of gene GO annotations\"\n",
    "PRECISION = \"precision\"\n",
    "RECALL = \"recall\"\n",
    "RECALL_GENERAL = \"recall_general\"\n",
    "RECALL_SPECIFIC = \"recall_specific\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "aa70e53d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from ontogpt.evaluation.enrichment.eval_enrichment import GeneSetComparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "87b881ca-c5a1-4e39-85ab-d3c5ba53b8d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_set</th>\n",
       "      <th>cutoff</th>\n",
       "      <th>closure</th>\n",
       "      <th>top_n</th>\n",
       "      <th>source</th>\n",
       "      <th>model</th>\n",
       "      <th>method</th>\n",
       "      <th>method_desc</th>\n",
       "      <th>run</th>\n",
       "      <th>truncation_factor</th>\n",
       "      <th>...</th>\n",
       "      <th>all_predictions_closure</th>\n",
       "      <th>unparsed</th>\n",
       "      <th>true_positive_terms</th>\n",
       "      <th>false_positive_terms</th>\n",
       "      <th>unparsed_terms</th>\n",
       "      <th>gene_set_size</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>recall_general</th>\n",
       "      <th>recall_specific</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>glycolysis-gocam-0-0.005</th>\n",
       "      <td>glycolysis-gocam-0</td>\n",
       "      <td>0.005</td>\n",
       "      <td>False</td>\n",
       "      <td>5</td>\n",
       "      <td>NONE</td>\n",
       "      <td>4.0</td>\n",
       "      <td>gpt</td>\n",
       "      <td>no_synopsis-4</td>\n",
       "      <td>gpt-4.no_synopsis.v1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "      <td>GO:0006096</td>\n",
       "      <td>GO:0006006</td>\n",
       "      <td>energy production|atp generation</td>\n",
       "      <td>10</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>glycolysis-gocam-0-0.005</th>\n",
       "      <td>glycolysis-gocam-0</td>\n",
       "      <td>0.005</td>\n",
       "      <td>False</td>\n",
       "      <td>10</td>\n",
       "      <td>NONE</td>\n",
       "      <td>4.0</td>\n",
       "      <td>gpt</td>\n",
       "      <td>no_synopsis-4</td>\n",
       "      <td>gpt-4.no_synopsis.v1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "      <td>GO:0006096</td>\n",
       "      <td>GO:0006006</td>\n",
       "      <td>energy production|atp generation</td>\n",
       "      <td>10</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>glycolysis-gocam-0-0.005</th>\n",
       "      <td>glycolysis-gocam-0</td>\n",
       "      <td>0.005</td>\n",
       "      <td>False</td>\n",
       "      <td>25</td>\n",
       "      <td>NONE</td>\n",
       "      <td>4.0</td>\n",
       "      <td>gpt</td>\n",
       "      <td>no_synopsis-4</td>\n",
       "      <td>gpt-4.no_synopsis.v1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "      <td>GO:0006096|GO:0006006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>energy production|atp generation</td>\n",
       "      <td>10</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.153846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>glycolysis-gocam-0-0.005</th>\n",
       "      <td>glycolysis-gocam-0</td>\n",
       "      <td>0.005</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>NONE</td>\n",
       "      <td>4.0</td>\n",
       "      <td>gpt</td>\n",
       "      <td>no_synopsis-4</td>\n",
       "      <td>gpt-4.no_synopsis.v1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "      <td>GO:0006096|GO:0006006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>energy production|atp generation</td>\n",
       "      <td>10</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.142857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>glycolysis-gocam-0-0.005</th>\n",
       "      <td>glycolysis-gocam-0</td>\n",
       "      <td>0.005</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>NONE</td>\n",
       "      <td>4.0</td>\n",
       "      <td>gpt</td>\n",
       "      <td>no_synopsis-4</td>\n",
       "      <td>gpt-4.no_synopsis.v1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "      <td>GO:0006096|GO:0019318|GO:0061621|GO:0006094</td>\n",
       "      <td>NaN</td>\n",
       "      <td>energy production|atp generation</td>\n",
       "      <td>10</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HALLMARK_MTORC1_SIGNALING-1-99</th>\n",
       "      <td>HALLMARK_MTORC1_SIGNALING-1</td>\n",
       "      <td>99.000</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>closure</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>4852</td>\n",
       "      <td>0</td>\n",
       "      <td>GO:0044283|GO:0044281|GO:0005737|GO:0019752|GO...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>180</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HALLMARK_MTORC1_SIGNALING-1-99</th>\n",
       "      <td>HALLMARK_MTORC1_SIGNALING-1</td>\n",
       "      <td>99.000</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>closure</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>4852</td>\n",
       "      <td>0</td>\n",
       "      <td>GO:0044283|GO:0005737|GO:0019752|GO:0036094|GO...</td>\n",
       "      <td>GO:0008525|GO:0009749|GO:0019899|GO:0050709|GO...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>180</td>\n",
       "      <td>0.000794</td>\n",
       "      <td>0.000794</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HALLMARK_MTORC1_SIGNALING-1-99</th>\n",
       "      <td>HALLMARK_MTORC1_SIGNALING-1</td>\n",
       "      <td>99.000</td>\n",
       "      <td>True</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>closure</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>4852</td>\n",
       "      <td>0</td>\n",
       "      <td>GO:0044283|GO:0005737|GO:0019752|GO:0036094|GO...</td>\n",
       "      <td>GO:0008525|GO:0009749|GO:0019899|GO:0050709|GO...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>180</td>\n",
       "      <td>0.001720</td>\n",
       "      <td>0.001720</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HALLMARK_MTORC1_SIGNALING-1-99</th>\n",
       "      <td>HALLMARK_MTORC1_SIGNALING-1</td>\n",
       "      <td>99.000</td>\n",
       "      <td>True</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>closure</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>4852</td>\n",
       "      <td>0</td>\n",
       "      <td>GO:0044283|GO:0005737|GO:0019752|GO:0036094|GO...</td>\n",
       "      <td>GO:0008525|GO:0019899|GO:0050709|GO:0035094|GO...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>180</td>\n",
       "      <td>0.004761</td>\n",
       "      <td>0.004761</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HALLMARK_MTORC1_SIGNALING-1-99</th>\n",
       "      <td>HALLMARK_MTORC1_SIGNALING-1</td>\n",
       "      <td>99.000</td>\n",
       "      <td>True</td>\n",
       "      <td>5000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>closure</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>4852</td>\n",
       "      <td>0</td>\n",
       "      <td>GO:0044283|GO:0005737|GO:0019752|GO:0036094|GO...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>180</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>86112 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   gene_set  cutoff  closure  \\\n",
       "name                                                                           \n",
       "glycolysis-gocam-0-0.005                 glycolysis-gocam-0   0.005    False   \n",
       "glycolysis-gocam-0-0.005                 glycolysis-gocam-0   0.005    False   \n",
       "glycolysis-gocam-0-0.005                 glycolysis-gocam-0   0.005    False   \n",
       "glycolysis-gocam-0-0.005                 glycolysis-gocam-0   0.005    False   \n",
       "glycolysis-gocam-0-0.005                 glycolysis-gocam-0   0.005     True   \n",
       "...                                                     ...     ...      ...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99  HALLMARK_MTORC1_SIGNALING-1  99.000    False   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99  HALLMARK_MTORC1_SIGNALING-1  99.000     True   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99  HALLMARK_MTORC1_SIGNALING-1  99.000     True   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99  HALLMARK_MTORC1_SIGNALING-1  99.000     True   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99  HALLMARK_MTORC1_SIGNALING-1  99.000     True   \n",
       "\n",
       "                                top_n source  model method    method_desc  \\\n",
       "name                                                                        \n",
       "glycolysis-gocam-0-0.005            5   NONE    4.0    gpt  no_synopsis-4   \n",
       "glycolysis-gocam-0-0.005           10   NONE    4.0    gpt  no_synopsis-4   \n",
       "glycolysis-gocam-0-0.005           25   NONE    4.0    gpt  no_synopsis-4   \n",
       "glycolysis-gocam-0-0.005         5000   NONE    4.0    gpt  no_synopsis-4   \n",
       "glycolysis-gocam-0-0.005            5   NONE    4.0    gpt  no_synopsis-4   \n",
       "...                               ...    ...    ...    ...            ...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99   5000    NaN    NaN    NaN            NaN   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99      5    NaN    NaN    NaN            NaN   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99     10    NaN    NaN    NaN            NaN   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99     25    NaN    NaN    NaN            NaN   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99   5000    NaN    NaN    NaN            NaN   \n",
       "\n",
       "                                                 run  truncation_factor  ...  \\\n",
       "name                                                                     ...   \n",
       "glycolysis-gocam-0-0.005        gpt-4.no_synopsis.v1                1.0  ...   \n",
       "glycolysis-gocam-0-0.005        gpt-4.no_synopsis.v1                1.0  ...   \n",
       "glycolysis-gocam-0-0.005        gpt-4.no_synopsis.v1                1.0  ...   \n",
       "glycolysis-gocam-0-0.005        gpt-4.no_synopsis.v1                1.0  ...   \n",
       "glycolysis-gocam-0-0.005        gpt-4.no_synopsis.v1                1.0  ...   \n",
       "...                                              ...                ...  ...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99               closure                NaN  ...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99               closure                NaN  ...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99               closure                NaN  ...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99               closure                NaN  ...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99               closure                NaN  ...   \n",
       "\n",
       "                               all_predictions_closure  unparsed  \\\n",
       "name                                                               \n",
       "glycolysis-gocam-0-0.005                            23         2   \n",
       "glycolysis-gocam-0-0.005                            23         2   \n",
       "glycolysis-gocam-0-0.005                            23         2   \n",
       "glycolysis-gocam-0-0.005                            23         2   \n",
       "glycolysis-gocam-0-0.005                            23         2   \n",
       "...                                                ...       ...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99                    4852         0   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99                    4852         0   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99                    4852         0   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99                    4852         0   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99                    4852         0   \n",
       "\n",
       "                                                              true_positive_terms  \\\n",
       "name                                                                                \n",
       "glycolysis-gocam-0-0.005                                               GO:0006096   \n",
       "glycolysis-gocam-0-0.005                                               GO:0006096   \n",
       "glycolysis-gocam-0-0.005                                    GO:0006096|GO:0006006   \n",
       "glycolysis-gocam-0-0.005                                    GO:0006096|GO:0006006   \n",
       "glycolysis-gocam-0-0.005              GO:0006096|GO:0019318|GO:0061621|GO:0006094   \n",
       "...                                                                           ...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99  GO:0044283|GO:0044281|GO:0005737|GO:0019752|GO...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99  GO:0044283|GO:0005737|GO:0019752|GO:0036094|GO...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99  GO:0044283|GO:0005737|GO:0019752|GO:0036094|GO...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99  GO:0044283|GO:0005737|GO:0019752|GO:0036094|GO...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99  GO:0044283|GO:0005737|GO:0019752|GO:0036094|GO...   \n",
       "\n",
       "                                                             false_positive_terms  \\\n",
       "name                                                                                \n",
       "glycolysis-gocam-0-0.005                                               GO:0006006   \n",
       "glycolysis-gocam-0-0.005                                               GO:0006006   \n",
       "glycolysis-gocam-0-0.005                                                      NaN   \n",
       "glycolysis-gocam-0-0.005                                                      NaN   \n",
       "glycolysis-gocam-0-0.005                                                      NaN   \n",
       "...                                                                           ...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99                                                NaN   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99  GO:0008525|GO:0009749|GO:0019899|GO:0050709|GO...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99  GO:0008525|GO:0009749|GO:0019899|GO:0050709|GO...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99  GO:0008525|GO:0019899|GO:0050709|GO:0035094|GO...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99                                                NaN   \n",
       "\n",
       "                                                  unparsed_terms  \\\n",
       "name                                                               \n",
       "glycolysis-gocam-0-0.005        energy production|atp generation   \n",
       "glycolysis-gocam-0-0.005        energy production|atp generation   \n",
       "glycolysis-gocam-0-0.005        energy production|atp generation   \n",
       "glycolysis-gocam-0-0.005        energy production|atp generation   \n",
       "glycolysis-gocam-0-0.005        energy production|atp generation   \n",
       "...                                                          ...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99                               NaN   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99                               NaN   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99                               NaN   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99                               NaN   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99                               NaN   \n",
       "\n",
       "                                gene_set_size  precision    recall  \\\n",
       "name                                                                 \n",
       "glycolysis-gocam-0-0.005                   10   0.500000  0.500000   \n",
       "glycolysis-gocam-0-0.005                   10   0.500000  0.500000   \n",
       "glycolysis-gocam-0-0.005                   10   1.000000  1.000000   \n",
       "glycolysis-gocam-0-0.005                   10   1.000000  1.000000   \n",
       "glycolysis-gocam-0-0.005                   10   1.000000  1.000000   \n",
       "...                                       ...        ...       ...   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99            180   1.000000  1.000000   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99            180   0.000794  0.000794   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99            180   0.001720  0.001720   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99            180   0.004761  0.004761   \n",
       "HALLMARK_MTORC1_SIGNALING-1-99            180   1.000000  1.000000   \n",
       "\n",
       "                                recall_general  recall_specific  \n",
       "name                                                             \n",
       "glycolysis-gocam-0-0.005              1.000000         0.200000  \n",
       "glycolysis-gocam-0-0.005              0.200000         0.200000  \n",
       "glycolysis-gocam-0-0.005              0.250000         0.153846  \n",
       "glycolysis-gocam-0-0.005              0.250000         0.142857  \n",
       "glycolysis-gocam-0-0.005              0.666667         0.800000  \n",
       "...                                        ...              ...  \n",
       "HALLMARK_MTORC1_SIGNALING-1-99        1.000000         1.000000  \n",
       "HALLMARK_MTORC1_SIGNALING-1-99        1.000000         1.000000  \n",
       "HALLMARK_MTORC1_SIGNALING-1-99        1.000000         1.000000  \n",
       "HALLMARK_MTORC1_SIGNALING-1-99        1.000000         1.000000  \n",
       "HALLMARK_MTORC1_SIGNALING-1-99        1.000000         1.000000  \n",
       "\n",
       "[86112 rows x 28 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('results/processed.tsv', sep='\\t', header=0, index_col=0)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "46f811ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "model\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>method</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>glycolysis-gocam-0-0.005</th>\n",
       "      <td>4.0</td>\n",
       "      <td>gpt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>glycolysis-gocam-0-0.005</th>\n",
       "      <td>3.5</td>\n",
       "      <td>gpt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>glycolysis-gocam-0-0.005</th>\n",
       "      <td>3.0</td>\n",
       "      <td>gpt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>glycolysis-gocam-0-0.005</th>\n",
       "      <td>NaN</td>\n",
       "      <td>standard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>glycolysis-gocam-0-0.005</th>\n",
       "      <td>NaN</td>\n",
       "      <td>standard_no_ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>glycolysis-gocam-0-0.005</th>\n",
       "      <td>NaN</td>\n",
       "      <td>random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>glycolysis-gocam-0-0.005</th>\n",
       "      <td>NaN</td>\n",
       "      <td>rank_based</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>glycolysis-gocam-0-0.005</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          model                method\n",
       "name                                                 \n",
       "glycolysis-gocam-0-0.005    4.0                   gpt\n",
       "glycolysis-gocam-0-0.005    3.5                   gpt\n",
       "glycolysis-gocam-0-0.005    3.0                   gpt\n",
       "glycolysis-gocam-0-0.005    NaN              standard\n",
       "glycolysis-gocam-0-0.005    NaN  standard_no_ontology\n",
       "glycolysis-gocam-0-0.005    NaN                random\n",
       "glycolysis-gocam-0-0.005    NaN            rank_based\n",
       "glycolysis-gocam-0-0.005    NaN                   NaN"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(MODEL)\n",
    "df[[MODEL, METHOD]].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "70226f4d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>glycolysis-gocam-0-0.005</th>\n",
       "      <td>glycolysis-gocam-0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>glycolysis-gocam-1-0.005</th>\n",
       "      <td>glycolysis-gocam-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bicluster_RNAseqDB_1001-0-0.005</th>\n",
       "      <td>bicluster_RNAseqDB_1001-0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bicluster_RNAseqDB_1001-1-0.005</th>\n",
       "      <td>bicluster_RNAseqDB_1001-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HALLMARK_HYPOXIA-0-0.005</th>\n",
       "      <td>HALLMARK_HYPOXIA-0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T cell proliferation-1-0.005</th>\n",
       "      <td>T cell proliferation-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mtorc1-0-0.005</th>\n",
       "      <td>mtorc1-0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mtorc1-1-0.005</th>\n",
       "      <td>mtorc1-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HALLMARK_MTORC1_SIGNALING-0-0.005</th>\n",
       "      <td>HALLMARK_MTORC1_SIGNALING-0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HALLMARK_MTORC1_SIGNALING-1-0.005</th>\n",
       "      <td>HALLMARK_MTORC1_SIGNALING-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>144 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      gene_set\n",
       "name                                                          \n",
       "glycolysis-gocam-0-0.005                    glycolysis-gocam-0\n",
       "glycolysis-gocam-1-0.005                    glycolysis-gocam-1\n",
       "bicluster_RNAseqDB_1001-0-0.005      bicluster_RNAseqDB_1001-0\n",
       "bicluster_RNAseqDB_1001-1-0.005      bicluster_RNAseqDB_1001-1\n",
       "HALLMARK_HYPOXIA-0-0.005                    HALLMARK_HYPOXIA-0\n",
       "...                                                        ...\n",
       "T cell proliferation-1-0.005            T cell proliferation-1\n",
       "mtorc1-0-0.005                                        mtorc1-0\n",
       "mtorc1-1-0.005                                        mtorc1-1\n",
       "HALLMARK_MTORC1_SIGNALING-0-0.005  HALLMARK_MTORC1_SIGNALING-0\n",
       "HALLMARK_MTORC1_SIGNALING-1-0.005  HALLMARK_MTORC1_SIGNALING-1\n",
       "\n",
       "[144 rows x 1 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[[GENESET]].drop_duplicates()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e937c8ba",
   "metadata": {},
   "source": [
    "## TABLE: All gene sets and their sizes\n",
    "\n",
    "Copy this to [gene set](https://docs.google.com/spreadsheets/d/1gGO5IHEg-N0hivtHBO6-rdXtin8hPhw-zv6eYOBgXcE/edit#gid=1762479413) tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "dd98fb6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "</style>\n",
       "<table id=\"T_fc4bd\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th id=\"T_fc4bd_level0_col0\" class=\"col_heading level0 col0\" >gene_set</th>\n",
       "      <th id=\"T_fc4bd_level0_col1\" class=\"col_heading level0 col1\" >gene_set_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row0_col0\" class=\"data row0 col0\" >glycolysis-gocam-0</td>\n",
       "      <td id=\"T_fc4bd_row0_col1\" class=\"data row0 col1\" >10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row1_col0\" class=\"data row1 col0\" >glycolysis-gocam-1</td>\n",
       "      <td id=\"T_fc4bd_row1_col1\" class=\"data row1 col1\" >9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row2_col0\" class=\"data row2 col0\" >bicluster_RNAseqDB_1001-0</td>\n",
       "      <td id=\"T_fc4bd_row2_col1\" class=\"data row2 col1\" >76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row3_col0\" class=\"data row3 col0\" >bicluster_RNAseqDB_1001-1</td>\n",
       "      <td id=\"T_fc4bd_row3_col1\" class=\"data row3 col1\" >63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row4_col0\" class=\"data row4 col0\" >HALLMARK_HYPOXIA-0</td>\n",
       "      <td id=\"T_fc4bd_row4_col1\" class=\"data row4 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row5_col0\" class=\"data row5 col0\" >HALLMARK_HYPOXIA-1</td>\n",
       "      <td id=\"T_fc4bd_row5_col1\" class=\"data row5 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row6_col0\" class=\"data row6 col0\" >HALLMARK_DNA_REPAIR-0</td>\n",
       "      <td id=\"T_fc4bd_row6_col1\" class=\"data row6 col1\" >150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row7_col0\" class=\"data row7 col0\" >HALLMARK_DNA_REPAIR-1</td>\n",
       "      <td id=\"T_fc4bd_row7_col1\" class=\"data row7 col1\" >135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row8_col0\" class=\"data row8 col0\" >HALLMARK_G2M_CHECKPOINT-0</td>\n",
       "      <td id=\"T_fc4bd_row8_col1\" class=\"data row8 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row9_col0\" class=\"data row9 col0\" >HALLMARK_G2M_CHECKPOINT-1</td>\n",
       "      <td id=\"T_fc4bd_row9_col1\" class=\"data row9 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row10_col0\" class=\"data row10 col0\" >EDS-0</td>\n",
       "      <td id=\"T_fc4bd_row10_col1\" class=\"data row10 col1\" >19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row11_col0\" class=\"data row11 col0\" >EDS-1</td>\n",
       "      <td id=\"T_fc4bd_row11_col1\" class=\"data row11 col1\" >18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row12_col0\" class=\"data row12 col0\" >HALLMARK_IL2_STAT5_SIGNALING-0</td>\n",
       "      <td id=\"T_fc4bd_row12_col1\" class=\"data row12 col1\" >199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row13_col0\" class=\"data row13 col0\" >HALLMARK_IL2_STAT5_SIGNALING-1</td>\n",
       "      <td id=\"T_fc4bd_row13_col1\" class=\"data row13 col1\" >179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row14_col0\" class=\"data row14 col0\" >HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION-0</td>\n",
       "      <td id=\"T_fc4bd_row14_col1\" class=\"data row14 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row15_col0\" class=\"data row15 col0\" >HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION-1</td>\n",
       "      <td id=\"T_fc4bd_row15_col1\" class=\"data row15 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row16_col0\" class=\"data row16 col0\" >HALLMARK_PI3K_AKT_MTOR_SIGNALING-0</td>\n",
       "      <td id=\"T_fc4bd_row16_col1\" class=\"data row16 col1\" >105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row17_col0\" class=\"data row17 col0\" >HALLMARK_PI3K_AKT_MTOR_SIGNALING-1</td>\n",
       "      <td id=\"T_fc4bd_row17_col1\" class=\"data row17 col1\" >95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row18_col0\" class=\"data row18 col0\" >HALLMARK_COAGULATION-0</td>\n",
       "      <td id=\"T_fc4bd_row18_col1\" class=\"data row18 col1\" >138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row19_col0\" class=\"data row19 col0\" >HALLMARK_COAGULATION-1</td>\n",
       "      <td id=\"T_fc4bd_row19_col1\" class=\"data row19 col1\" >125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row20_col0\" class=\"data row20 col0\" >peroxisome-0</td>\n",
       "      <td id=\"T_fc4bd_row20_col1\" class=\"data row20 col1\" >8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row21_col0\" class=\"data row21 col0\" >peroxisome-1</td>\n",
       "      <td id=\"T_fc4bd_row21_col1\" class=\"data row21 col1\" >5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row22_col0\" class=\"data row22 col0\" >HALLMARK_APICAL_JUNCTION-0</td>\n",
       "      <td id=\"T_fc4bd_row22_col1\" class=\"data row22 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row23_col0\" class=\"data row23 col0\" >HALLMARK_APICAL_JUNCTION-1</td>\n",
       "      <td id=\"T_fc4bd_row23_col1\" class=\"data row23 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row24_col0\" class=\"data row24 col0\" >HALLMARK_ANGIOGENESIS-0</td>\n",
       "      <td id=\"T_fc4bd_row24_col1\" class=\"data row24 col1\" >36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row25_col0\" class=\"data row25 col0\" >HALLMARK_ANGIOGENESIS-1</td>\n",
       "      <td id=\"T_fc4bd_row25_col1\" class=\"data row25 col1\" >33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row26_col0\" class=\"data row26 col0\" >go-postsynapse-calcium-transmembrane-0</td>\n",
       "      <td id=\"T_fc4bd_row26_col1\" class=\"data row26 col1\" >33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row27_col0\" class=\"data row27 col0\" >go-postsynapse-calcium-transmembrane-1</td>\n",
       "      <td id=\"T_fc4bd_row27_col1\" class=\"data row27 col1\" >30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row28_col0\" class=\"data row28 col0\" >HALLMARK_BILE_ACID_METABOLISM-0</td>\n",
       "      <td id=\"T_fc4bd_row28_col1\" class=\"data row28 col1\" >112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row29_col0\" class=\"data row29 col0\" >HALLMARK_BILE_ACID_METABOLISM-1</td>\n",
       "      <td id=\"T_fc4bd_row29_col1\" class=\"data row29 col1\" >101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row30_col0\" class=\"data row30 col0\" >HALLMARK_CHOLESTEROL_HOMEOSTASIS-0</td>\n",
       "      <td id=\"T_fc4bd_row30_col1\" class=\"data row30 col1\" >74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row31_col0\" class=\"data row31 col0\" >HALLMARK_CHOLESTEROL_HOMEOSTASIS-1</td>\n",
       "      <td id=\"T_fc4bd_row31_col1\" class=\"data row31 col1\" >67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row32_col0\" class=\"data row32 col0\" >bicluster_RNAseqDB_1002-0</td>\n",
       "      <td id=\"T_fc4bd_row32_col1\" class=\"data row32 col1\" >52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row33_col0\" class=\"data row33 col0\" >bicluster_RNAseqDB_1002-1</td>\n",
       "      <td id=\"T_fc4bd_row33_col1\" class=\"data row33 col1\" >43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row34_col0\" class=\"data row34 col0\" >HALLMARK_KRAS_SIGNALING_DN-0</td>\n",
       "      <td id=\"T_fc4bd_row34_col1\" class=\"data row34 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row35_col0\" class=\"data row35 col0\" >HALLMARK_KRAS_SIGNALING_DN-1</td>\n",
       "      <td id=\"T_fc4bd_row35_col1\" class=\"data row35 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row36_col0\" class=\"data row36 col0\" >bicluster_RNAseqDB_0-0</td>\n",
       "      <td id=\"T_fc4bd_row36_col1\" class=\"data row36 col1\" >158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row37_col0\" class=\"data row37 col0\" >bicluster_RNAseqDB_0-1</td>\n",
       "      <td id=\"T_fc4bd_row37_col1\" class=\"data row37 col1\" >134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row38_col0\" class=\"data row38 col0\" >amigo-example-0</td>\n",
       "      <td id=\"T_fc4bd_row38_col1\" class=\"data row38 col1\" >36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row39_col0\" class=\"data row39 col0\" >amigo-example-1</td>\n",
       "      <td id=\"T_fc4bd_row39_col1\" class=\"data row39 col1\" >32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row40_col0\" class=\"data row40 col0\" >HALLMARK_MYOGENESIS-0</td>\n",
       "      <td id=\"T_fc4bd_row40_col1\" class=\"data row40 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row41_col0\" class=\"data row41 col0\" >HALLMARK_MYOGENESIS-1</td>\n",
       "      <td id=\"T_fc4bd_row41_col1\" class=\"data row41 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row42_col0\" class=\"data row42 col0\" >HALLMARK_MITOTIC_SPINDLE-0</td>\n",
       "      <td id=\"T_fc4bd_row42_col1\" class=\"data row42 col1\" >199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row43_col0\" class=\"data row43 col0\" >HALLMARK_MITOTIC_SPINDLE-1</td>\n",
       "      <td id=\"T_fc4bd_row43_col1\" class=\"data row43 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row44_col0\" class=\"data row44 col0\" >HALLMARK_KRAS_SIGNALING_UP-0</td>\n",
       "      <td id=\"T_fc4bd_row44_col1\" class=\"data row44 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row45_col0\" class=\"data row45 col0\" >HALLMARK_KRAS_SIGNALING_UP-1</td>\n",
       "      <td id=\"T_fc4bd_row45_col1\" class=\"data row45 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row46_col0\" class=\"data row46 col0\" >HALLMARK_GLYCOLYSIS-0</td>\n",
       "      <td id=\"T_fc4bd_row46_col1\" class=\"data row46 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row47_col0\" class=\"data row47 col0\" >HALLMARK_GLYCOLYSIS-1</td>\n",
       "      <td id=\"T_fc4bd_row47_col1\" class=\"data row47 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row48_col0\" class=\"data row48 col0\" >term-GO:0007212-0</td>\n",
       "      <td id=\"T_fc4bd_row48_col1\" class=\"data row48 col1\" >28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row49_col0\" class=\"data row49 col0\" >term-GO:0007212-1</td>\n",
       "      <td id=\"T_fc4bd_row49_col1\" class=\"data row49 col1\" >26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row50_col0\" class=\"data row50 col0\" >HALLMARK_ANDROGEN_RESPONSE-0</td>\n",
       "      <td id=\"T_fc4bd_row50_col1\" class=\"data row50 col1\" >101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row51_col0\" class=\"data row51 col0\" >HALLMARK_ANDROGEN_RESPONSE-1</td>\n",
       "      <td id=\"T_fc4bd_row51_col1\" class=\"data row51 col1\" >90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row52_col0\" class=\"data row52 col0\" >HALLMARK_ESTROGEN_RESPONSE_EARLY-0</td>\n",
       "      <td id=\"T_fc4bd_row52_col1\" class=\"data row52 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row53_col0\" class=\"data row53 col0\" >HALLMARK_ESTROGEN_RESPONSE_EARLY-1</td>\n",
       "      <td id=\"T_fc4bd_row53_col1\" class=\"data row53 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row54_col0\" class=\"data row54 col0\" >endocytosis-0</td>\n",
       "      <td id=\"T_fc4bd_row54_col1\" class=\"data row54 col1\" >16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row55_col0\" class=\"data row55 col0\" >endocytosis-1</td>\n",
       "      <td id=\"T_fc4bd_row55_col1\" class=\"data row55 col1\" >15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row56_col0\" class=\"data row56 col0\" >regulation of presynaptic membrane potential-0</td>\n",
       "      <td id=\"T_fc4bd_row56_col1\" class=\"data row56 col1\" >30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row57_col0\" class=\"data row57 col0\" >regulation of presynaptic membrane potential-1</td>\n",
       "      <td id=\"T_fc4bd_row57_col1\" class=\"data row57 col1\" >27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row58_col0\" class=\"data row58 col0\" >HALLMARK_TNFA_SIGNALING_VIA_NFKB-0</td>\n",
       "      <td id=\"T_fc4bd_row58_col1\" class=\"data row58 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row59_col0\" class=\"data row59 col0\" >HALLMARK_TNFA_SIGNALING_VIA_NFKB-1</td>\n",
       "      <td id=\"T_fc4bd_row59_col1\" class=\"data row59 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row60_col0\" class=\"data row60 col0\" >hydrolase activity, hydrolyzing O-glycosyl compounds-0</td>\n",
       "      <td id=\"T_fc4bd_row60_col1\" class=\"data row60 col1\" >91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row61_col0\" class=\"data row61 col0\" >hydrolase activity, hydrolyzing O-glycosyl compounds-1</td>\n",
       "      <td id=\"T_fc4bd_row61_col1\" class=\"data row61 col1\" >81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row62_col0\" class=\"data row62 col0\" >HALLMARK_UNFOLDED_PROTEIN_RESPONSE-0</td>\n",
       "      <td id=\"T_fc4bd_row62_col1\" class=\"data row62 col1\" >113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row63_col0\" class=\"data row63 col0\" >HALLMARK_UNFOLDED_PROTEIN_RESPONSE-1</td>\n",
       "      <td id=\"T_fc4bd_row63_col1\" class=\"data row63 col1\" >101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row64_col0\" class=\"data row64 col0\" >molecular sequestering-0</td>\n",
       "      <td id=\"T_fc4bd_row64_col1\" class=\"data row64 col1\" >30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row65_col0\" class=\"data row65 col0\" >molecular sequestering-1</td>\n",
       "      <td id=\"T_fc4bd_row65_col1\" class=\"data row65 col1\" >27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row66_col0\" class=\"data row66 col0\" >tf-downreg-colorectal-0</td>\n",
       "      <td id=\"T_fc4bd_row66_col1\" class=\"data row66 col1\" >51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row67_col0\" class=\"data row67 col0\" >tf-downreg-colorectal-1</td>\n",
       "      <td id=\"T_fc4bd_row67_col1\" class=\"data row67 col1\" >46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row68_col0\" class=\"data row68 col0\" >HALLMARK_P53_PATHWAY-0</td>\n",
       "      <td id=\"T_fc4bd_row68_col1\" class=\"data row68 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row69_col0\" class=\"data row69 col0\" >HALLMARK_P53_PATHWAY-1</td>\n",
       "      <td id=\"T_fc4bd_row69_col1\" class=\"data row69 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row70_col0\" class=\"data row70 col0\" >HALLMARK_OXIDATIVE_PHOSPHORYLATION-0</td>\n",
       "      <td id=\"T_fc4bd_row70_col1\" class=\"data row70 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row71_col0\" class=\"data row71 col0\" >HALLMARK_OXIDATIVE_PHOSPHORYLATION-1</td>\n",
       "      <td id=\"T_fc4bd_row71_col1\" class=\"data row71 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row72_col0\" class=\"data row72 col0\" >FA-0</td>\n",
       "      <td id=\"T_fc4bd_row72_col1\" class=\"data row72 col1\" >19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row73_col0\" class=\"data row73 col0\" >FA-1</td>\n",
       "      <td id=\"T_fc4bd_row73_col1\" class=\"data row73 col1\" >18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row74_col0\" class=\"data row74 col0\" >HALLMARK_NOTCH_SIGNALING-0</td>\n",
       "      <td id=\"T_fc4bd_row74_col1\" class=\"data row74 col1\" >32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row75_col0\" class=\"data row75 col0\" >HALLMARK_NOTCH_SIGNALING-1</td>\n",
       "      <td id=\"T_fc4bd_row75_col1\" class=\"data row75 col1\" >29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row76_col0\" class=\"data row76 col0\" >HALLMARK_PROTEIN_SECRETION-0</td>\n",
       "      <td id=\"T_fc4bd_row76_col1\" class=\"data row76 col1\" >96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row77_col0\" class=\"data row77 col0\" >HALLMARK_PROTEIN_SECRETION-1</td>\n",
       "      <td id=\"T_fc4bd_row77_col1\" class=\"data row77 col1\" >87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row78_col0\" class=\"data row78 col0\" >HALLMARK_FATTY_ACID_METABOLISM-0</td>\n",
       "      <td id=\"T_fc4bd_row78_col1\" class=\"data row78 col1\" >158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row79_col0\" class=\"data row79 col0\" >HALLMARK_FATTY_ACID_METABOLISM-1</td>\n",
       "      <td id=\"T_fc4bd_row79_col1\" class=\"data row79 col1\" >143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row80_col0\" class=\"data row80 col0\" >ig-receptor-binding-2022-0</td>\n",
       "      <td id=\"T_fc4bd_row80_col1\" class=\"data row80 col1\" >91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row81_col0\" class=\"data row81 col0\" >ig-receptor-binding-2022-1</td>\n",
       "      <td id=\"T_fc4bd_row81_col1\" class=\"data row81 col1\" >82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row82_col0\" class=\"data row82 col0\" >HALLMARK_UV_RESPONSE_UP-0</td>\n",
       "      <td id=\"T_fc4bd_row82_col1\" class=\"data row82 col1\" >158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row83_col0\" class=\"data row83 col0\" >HALLMARK_UV_RESPONSE_UP-1</td>\n",
       "      <td id=\"T_fc4bd_row83_col1\" class=\"data row83 col1\" >143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row84_col0\" class=\"data row84 col0\" >HALLMARK_COMPLEMENT-0</td>\n",
       "      <td id=\"T_fc4bd_row84_col1\" class=\"data row84 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row85_col0\" class=\"data row85 col0\" >HALLMARK_COMPLEMENT-1</td>\n",
       "      <td id=\"T_fc4bd_row85_col1\" class=\"data row85 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row86_col0\" class=\"data row86 col0\" >HALLMARK_UV_RESPONSE_DN-0</td>\n",
       "      <td id=\"T_fc4bd_row86_col1\" class=\"data row86 col1\" >144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row87_col0\" class=\"data row87 col0\" >HALLMARK_UV_RESPONSE_DN-1</td>\n",
       "      <td id=\"T_fc4bd_row87_col1\" class=\"data row87 col1\" >130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row88_col0\" class=\"data row88 col0\" >HALLMARK_MYC_TARGETS_V1-0</td>\n",
       "      <td id=\"T_fc4bd_row88_col1\" class=\"data row88 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row89_col0\" class=\"data row89 col0\" >HALLMARK_MYC_TARGETS_V1-1</td>\n",
       "      <td id=\"T_fc4bd_row89_col1\" class=\"data row89 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row90_col0\" class=\"data row90 col0\" >HALLMARK_HEME_METABOLISM-0</td>\n",
       "      <td id=\"T_fc4bd_row90_col1\" class=\"data row90 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row91_col0\" class=\"data row91 col0\" >HALLMARK_HEME_METABOLISM-1</td>\n",
       "      <td id=\"T_fc4bd_row91_col1\" class=\"data row91 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row92_col0\" class=\"data row92 col0\" >HALLMARK_SPERMATOGENESIS-0</td>\n",
       "      <td id=\"T_fc4bd_row92_col1\" class=\"data row92 col1\" >135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row93_col0\" class=\"data row93 col0\" >HALLMARK_SPERMATOGENESIS-1</td>\n",
       "      <td id=\"T_fc4bd_row93_col1\" class=\"data row93 col1\" >122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row94_col0\" class=\"data row94 col0\" >HALLMARK_INTERFERON_GAMMA_RESPONSE-0</td>\n",
       "      <td id=\"T_fc4bd_row94_col1\" class=\"data row94 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row95_col0\" class=\"data row95 col0\" >HALLMARK_INTERFERON_GAMMA_RESPONSE-1</td>\n",
       "      <td id=\"T_fc4bd_row95_col1\" class=\"data row95 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row96_col0\" class=\"data row96 col0\" >HALLMARK_TGF_BETA_SIGNALING-0</td>\n",
       "      <td id=\"T_fc4bd_row96_col1\" class=\"data row96 col1\" >54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row97_col0\" class=\"data row97 col0\" >HALLMARK_TGF_BETA_SIGNALING-1</td>\n",
       "      <td id=\"T_fc4bd_row97_col1\" class=\"data row97 col1\" >49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row98_col0\" class=\"data row98 col0\" >HALLMARK_PEROXISOME-0</td>\n",
       "      <td id=\"T_fc4bd_row98_col1\" class=\"data row98 col1\" >104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row99_col0\" class=\"data row99 col0\" >HALLMARK_PEROXISOME-1</td>\n",
       "      <td id=\"T_fc4bd_row99_col1\" class=\"data row99 col1\" >94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row100_col0\" class=\"data row100 col0\" >HALLMARK_APOPTOSIS-0</td>\n",
       "      <td id=\"T_fc4bd_row100_col1\" class=\"data row100 col1\" >161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row101_col0\" class=\"data row101 col0\" >HALLMARK_APOPTOSIS-1</td>\n",
       "      <td id=\"T_fc4bd_row101_col1\" class=\"data row101 col1\" >145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row102_col0\" class=\"data row102 col0\" >sensory ataxia-0</td>\n",
       "      <td id=\"T_fc4bd_row102_col1\" class=\"data row102 col1\" >15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row103_col0\" class=\"data row103 col0\" >sensory ataxia-1</td>\n",
       "      <td id=\"T_fc4bd_row103_col1\" class=\"data row103 col1\" >14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row104_col0\" class=\"data row104 col0\" >HALLMARK_ADIPOGENESIS-0</td>\n",
       "      <td id=\"T_fc4bd_row104_col1\" class=\"data row104 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row105_col0\" class=\"data row105 col0\" >HALLMARK_ADIPOGENESIS-1</td>\n",
       "      <td id=\"T_fc4bd_row105_col1\" class=\"data row105 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row106_col0\" class=\"data row106 col0\" >HALLMARK_ALLOGRAFT_REJECTION-0</td>\n",
       "      <td id=\"T_fc4bd_row106_col1\" class=\"data row106 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row107_col0\" class=\"data row107 col0\" >HALLMARK_ALLOGRAFT_REJECTION-1</td>\n",
       "      <td id=\"T_fc4bd_row107_col1\" class=\"data row107 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row108_col0\" class=\"data row108 col0\" >HALLMARK_INFLAMMATORY_RESPONSE-0</td>\n",
       "      <td id=\"T_fc4bd_row108_col1\" class=\"data row108 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row109_col0\" class=\"data row109 col0\" >HALLMARK_INFLAMMATORY_RESPONSE-1</td>\n",
       "      <td id=\"T_fc4bd_row109_col1\" class=\"data row109 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row110_col0\" class=\"data row110 col0\" >HALLMARK_WNT_BETA_CATENIN_SIGNALING-0</td>\n",
       "      <td id=\"T_fc4bd_row110_col1\" class=\"data row110 col1\" >42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row111_col0\" class=\"data row111 col0\" >HALLMARK_WNT_BETA_CATENIN_SIGNALING-1</td>\n",
       "      <td id=\"T_fc4bd_row111_col1\" class=\"data row111 col1\" >38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row112_col0\" class=\"data row112 col0\" >HALLMARK_E2F_TARGETS-0</td>\n",
       "      <td id=\"T_fc4bd_row112_col1\" class=\"data row112 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row113_col0\" class=\"data row113 col0\" >HALLMARK_E2F_TARGETS-1</td>\n",
       "      <td id=\"T_fc4bd_row113_col1\" class=\"data row113 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row114_col0\" class=\"data row114 col0\" >meiosis I-0</td>\n",
       "      <td id=\"T_fc4bd_row114_col1\" class=\"data row114 col1\" >54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row115_col0\" class=\"data row115 col0\" >meiosis I-1</td>\n",
       "      <td id=\"T_fc4bd_row115_col1\" class=\"data row115 col1\" >46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row116_col0\" class=\"data row116 col0\" >HALLMARK_ESTROGEN_RESPONSE_LATE-0</td>\n",
       "      <td id=\"T_fc4bd_row116_col1\" class=\"data row116 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row117_col0\" class=\"data row117 col0\" >HALLMARK_ESTROGEN_RESPONSE_LATE-1</td>\n",
       "      <td id=\"T_fc4bd_row117_col1\" class=\"data row117 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row118_col0\" class=\"data row118 col0\" >progeria-0</td>\n",
       "      <td id=\"T_fc4bd_row118_col1\" class=\"data row118 col1\" >4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row119_col0\" class=\"data row119 col0\" >progeria-1</td>\n",
       "      <td id=\"T_fc4bd_row119_col1\" class=\"data row119 col1\" >3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row120_col0\" class=\"data row120 col0\" >HALLMARK_MYC_TARGETS_V2-0</td>\n",
       "      <td id=\"T_fc4bd_row120_col1\" class=\"data row120 col1\" >58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row121_col0\" class=\"data row121 col0\" >HALLMARK_MYC_TARGETS_V2-1</td>\n",
       "      <td id=\"T_fc4bd_row121_col1\" class=\"data row121 col1\" >53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row122_col0\" class=\"data row122 col0\" >Yamanaka-TFs-0</td>\n",
       "      <td id=\"T_fc4bd_row122_col1\" class=\"data row122 col1\" >4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row123_col0\" class=\"data row123 col0\" >Yamanaka-TFs-1</td>\n",
       "      <td id=\"T_fc4bd_row123_col1\" class=\"data row123 col1\" >3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row124_col0\" class=\"data row124 col0\" >HALLMARK_IL6_JAK_STAT3_SIGNALING-0</td>\n",
       "      <td id=\"T_fc4bd_row124_col1\" class=\"data row124 col1\" >87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row125_col0\" class=\"data row125 col0\" >HALLMARK_IL6_JAK_STAT3_SIGNALING-1</td>\n",
       "      <td id=\"T_fc4bd_row125_col1\" class=\"data row125 col1\" >79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row126_col0\" class=\"data row126 col0\" >HALLMARK_INTERFERON_ALPHA_RESPONSE-0</td>\n",
       "      <td id=\"T_fc4bd_row126_col1\" class=\"data row126 col1\" >97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row127_col0\" class=\"data row127 col0\" >HALLMARK_INTERFERON_ALPHA_RESPONSE-1</td>\n",
       "      <td id=\"T_fc4bd_row127_col1\" class=\"data row127 col1\" >88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row128_col0\" class=\"data row128 col0\" >HALLMARK_PANCREAS_BETA_CELLS-0</td>\n",
       "      <td id=\"T_fc4bd_row128_col1\" class=\"data row128 col1\" >40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row129_col0\" class=\"data row129 col0\" >HALLMARK_PANCREAS_BETA_CELLS-1</td>\n",
       "      <td id=\"T_fc4bd_row129_col1\" class=\"data row129 col1\" >36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row130_col0\" class=\"data row130 col0\" >go-reg-autophagy-pkra-0</td>\n",
       "      <td id=\"T_fc4bd_row130_col1\" class=\"data row130 col1\" >17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row131_col0\" class=\"data row131 col0\" >go-reg-autophagy-pkra-1</td>\n",
       "      <td id=\"T_fc4bd_row131_col1\" class=\"data row131 col1\" >16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row132_col0\" class=\"data row132 col0\" >HALLMARK_APICAL_SURFACE-0</td>\n",
       "      <td id=\"T_fc4bd_row132_col1\" class=\"data row132 col1\" >44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row133_col0\" class=\"data row133 col0\" >HALLMARK_APICAL_SURFACE-1</td>\n",
       "      <td id=\"T_fc4bd_row133_col1\" class=\"data row133 col1\" >40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row134_col0\" class=\"data row134 col0\" >HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY-0</td>\n",
       "      <td id=\"T_fc4bd_row134_col1\" class=\"data row134 col1\" >49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row135_col0\" class=\"data row135 col0\" >HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY-1</td>\n",
       "      <td id=\"T_fc4bd_row135_col1\" class=\"data row135 col1\" >45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row136_col0\" class=\"data row136 col0\" >HALLMARK_HEDGEHOG_SIGNALING-0</td>\n",
       "      <td id=\"T_fc4bd_row136_col1\" class=\"data row136 col1\" >36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row137_col0\" class=\"data row137 col0\" >HALLMARK_HEDGEHOG_SIGNALING-1</td>\n",
       "      <td id=\"T_fc4bd_row137_col1\" class=\"data row137 col1\" >33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row138_col0\" class=\"data row138 col0\" >T cell proliferation-0</td>\n",
       "      <td id=\"T_fc4bd_row138_col1\" class=\"data row138 col1\" >72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row139_col0\" class=\"data row139 col0\" >T cell proliferation-1</td>\n",
       "      <td id=\"T_fc4bd_row139_col1\" class=\"data row139 col1\" >65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row140_col0\" class=\"data row140 col0\" >mtorc1-0</td>\n",
       "      <td id=\"T_fc4bd_row140_col1\" class=\"data row140 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row141_col0\" class=\"data row141 col0\" >mtorc1-1</td>\n",
       "      <td id=\"T_fc4bd_row141_col1\" class=\"data row141 col1\" >180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row142_col0\" class=\"data row142 col0\" >HALLMARK_MTORC1_SIGNALING-0</td>\n",
       "      <td id=\"T_fc4bd_row142_col1\" class=\"data row142 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fc4bd_row143_col0\" class=\"data row143 col0\" >HALLMARK_MTORC1_SIGNALING-1</td>\n",
       "      <td id=\"T_fc4bd_row143_col1\" class=\"data row143 col1\" >180</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x10359dea0>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[[GENESET, GENESET_SIZE]].drop_duplicates().style.hide()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "26561796",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "</style>\n",
       "<table id=\"T_bbf05\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th id=\"T_bbf05_level0_col0\" class=\"col_heading level0 col0\" >model</th>\n",
       "      <th id=\"T_bbf05_level0_col1\" class=\"col_heading level0 col1\" >method</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_bbf05_row0_col0\" class=\"data row0 col0\" >4.000000</td>\n",
       "      <td id=\"T_bbf05_row0_col1\" class=\"data row0 col1\" >gpt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bbf05_row1_col0\" class=\"data row1 col0\" >3.500000</td>\n",
       "      <td id=\"T_bbf05_row1_col1\" class=\"data row1 col1\" >gpt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bbf05_row2_col0\" class=\"data row2 col0\" >3.000000</td>\n",
       "      <td id=\"T_bbf05_row2_col1\" class=\"data row2 col1\" >gpt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bbf05_row3_col0\" class=\"data row3 col0\" >nan</td>\n",
       "      <td id=\"T_bbf05_row3_col1\" class=\"data row3 col1\" >standard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bbf05_row4_col0\" class=\"data row4 col0\" >nan</td>\n",
       "      <td id=\"T_bbf05_row4_col1\" class=\"data row4 col1\" >standard_no_ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bbf05_row5_col0\" class=\"data row5 col0\" >nan</td>\n",
       "      <td id=\"T_bbf05_row5_col1\" class=\"data row5 col1\" >random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bbf05_row6_col0\" class=\"data row6 col0\" >nan</td>\n",
       "      <td id=\"T_bbf05_row6_col1\" class=\"data row6 col1\" >rank_based</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bbf05_row7_col0\" class=\"data row7 col0\" >nan</td>\n",
       "      <td id=\"T_bbf05_row7_col1\" class=\"data row7 col1\" >nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x103506ce0>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[[MODEL, METHOD]].drop_duplicates().style.hide()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "97f4f9e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "</style>\n",
       "<table id=\"T_e0d75\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th id=\"T_e0d75_level0_col0\" class=\"col_heading level0 col0\" >model</th>\n",
       "      <th id=\"T_e0d75_level0_col1\" class=\"col_heading level0 col1\" >method</th>\n",
       "      <th id=\"T_e0d75_level0_col2\" class=\"col_heading level0 col2\" >prompt_variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_e0d75_row0_col0\" class=\"data row0 col0\" >4.000000</td>\n",
       "      <td id=\"T_e0d75_row0_col1\" class=\"data row0 col1\" >gpt</td>\n",
       "      <td id=\"T_e0d75_row0_col2\" class=\"data row0 col2\" >v1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e0d75_row1_col0\" class=\"data row1 col0\" >4.000000</td>\n",
       "      <td id=\"T_e0d75_row1_col1\" class=\"data row1 col1\" >gpt</td>\n",
       "      <td id=\"T_e0d75_row1_col2\" class=\"data row1 col2\" >v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e0d75_row2_col0\" class=\"data row2 col0\" >3.500000</td>\n",
       "      <td id=\"T_e0d75_row2_col1\" class=\"data row2 col1\" >gpt</td>\n",
       "      <td id=\"T_e0d75_row2_col2\" class=\"data row2 col2\" >v1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e0d75_row3_col0\" class=\"data row3 col0\" >3.500000</td>\n",
       "      <td id=\"T_e0d75_row3_col1\" class=\"data row3 col1\" >gpt</td>\n",
       "      <td id=\"T_e0d75_row3_col2\" class=\"data row3 col2\" >v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e0d75_row4_col0\" class=\"data row4 col0\" >3.000000</td>\n",
       "      <td id=\"T_e0d75_row4_col1\" class=\"data row4 col1\" >gpt</td>\n",
       "      <td id=\"T_e0d75_row4_col2\" class=\"data row4 col2\" >v1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e0d75_row5_col0\" class=\"data row5 col0\" >3.000000</td>\n",
       "      <td id=\"T_e0d75_row5_col1\" class=\"data row5 col1\" >gpt</td>\n",
       "      <td id=\"T_e0d75_row5_col2\" class=\"data row5 col2\" >v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e0d75_row6_col0\" class=\"data row6 col0\" >nan</td>\n",
       "      <td id=\"T_e0d75_row6_col1\" class=\"data row6 col1\" >standard</td>\n",
       "      <td id=\"T_e0d75_row6_col2\" class=\"data row6 col2\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e0d75_row7_col0\" class=\"data row7 col0\" >nan</td>\n",
       "      <td id=\"T_e0d75_row7_col1\" class=\"data row7 col1\" >standard_no_ontology</td>\n",
       "      <td id=\"T_e0d75_row7_col2\" class=\"data row7 col2\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e0d75_row8_col0\" class=\"data row8 col0\" >nan</td>\n",
       "      <td id=\"T_e0d75_row8_col1\" class=\"data row8 col1\" >random</td>\n",
       "      <td id=\"T_e0d75_row8_col2\" class=\"data row8 col2\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e0d75_row9_col0\" class=\"data row9 col0\" >nan</td>\n",
       "      <td id=\"T_e0d75_row9_col1\" class=\"data row9 col1\" >rank_based</td>\n",
       "      <td id=\"T_e0d75_row9_col2\" class=\"data row9 col2\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e0d75_row10_col0\" class=\"data row10 col0\" >nan</td>\n",
       "      <td id=\"T_e0d75_row10_col1\" class=\"data row10 col1\" >nan</td>\n",
       "      <td id=\"T_e0d75_row10_col2\" class=\"data row10 col2\" >nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x296809c30>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[[MODEL, METHOD, PROMPT_VARIANT]].drop_duplicates().style.hide()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "881a059f-8548-40dd-bdb9-e1af7fd9bed3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_set</th>\n",
       "      <th>cutoff</th>\n",
       "      <th>closure</th>\n",
       "      <th>top_n</th>\n",
       "      <th>source</th>\n",
       "      <th>model</th>\n",
       "      <th>method</th>\n",
       "      <th>method_desc</th>\n",
       "      <th>run</th>\n",
       "      <th>truncation_factor</th>\n",
       "      <th>...</th>\n",
       "      <th>all_predictions_closure</th>\n",
       "      <th>unparsed</th>\n",
       "      <th>true_positive_terms</th>\n",
       "      <th>false_positive_terms</th>\n",
       "      <th>unparsed_terms</th>\n",
       "      <th>gene_set_size</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>recall_general</th>\n",
       "      <th>recall_specific</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene_set, cutoff, closure, top_n, source, model, method, method_desc, run, truncation_factor, prompt_variant, response_token_length, prompt_length, true_positives, false_positives, false_negatives, more_general_false_negatives, more_specific_false_negatives, all_predictions_closure, unparsed, true_positive_terms, false_positive_terms, unparsed_terms, gene_set_size, precision, recall, recall_general, recall_specific]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 28 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_orig = df.copy()\n",
    "\n",
    "df = df.loc[(df[\"cutoff\"] == 0.005) & (df[\"closure\"] == True) & (df[\"top_n\"] == 1)] \n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d51e6f2f",
   "metadata": {},
   "source": [
    "## Subset Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7808ed4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "go = get_adapter(\"sqlite:obo:go\")\n",
    "subsets = list(go.subsets())\n",
    "subsets_by_term = defaultdict(list)\n",
    "for subset in subsets:\n",
    "    members = [m for m in go.subset_members(subset) if m.startswith(\"GO:\")]\n",
    "    for t in members:\n",
    "        subsets_by_term[t].append(subset)\n",
    "    members_ancs = go.ancestors(members)\n",
    "    anc_subset = f\"anc_of_{subset}\"\n",
    "    for a in members_ancs:\n",
    "        if a not in members:\n",
    "            subsets_by_term[a].append(anc_subset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0767dadd",
   "metadata": {},
   "outputs": [],
   "source": [
    "go = get_adapter(\"sqlite:obo:go\")\n",
    "labels = list(go.labels(go.entities()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "48103d10",
   "metadata": {},
   "outputs": [],
   "source": [
    "lmap = {id: label for id, label in labels}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "bd1777b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "objs = []\n",
    "n = 0\n",
    "for row in df.to_dict(orient=\"records\"):\n",
    "    n += 1\n",
    "    if n % 500 == 0:\n",
    "        print(row[GENESET])\n",
    "    for t in row[GO_TERM_IDS]:\n",
    "        obj = {MODEL: row.get(MODEL), METHOD: row.get(METHOD) , \"term\": t, \"label\": lmap.get(t, t)}\n",
    "        for s in subsets_by_term.get(t, []):\n",
    "            obj[s] = 1\n",
    "        objs.append(obj)\n",
    "print(len(objs))\n",
    "subsets_df=pd.DataFrame(objs) \n",
    "pd.set_option('display.max_rows', 10)\n",
    "subsets_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55f35036",
   "metadata": {},
   "source": [
    "### All subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e7df46ee",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'model'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[19], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43msubsets_df\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfillna\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgroupby\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43mMODEL\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mMETHOD\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mmean(numeric_only\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\u001b[38;5;241m.\u001b[39mstyle\u001b[38;5;241m.\u001b[39mhighlight_max(axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m, props\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfont-weight:bold\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;241m.\u001b[39mformat(precision\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m3\u001b[39m)\n",
      "File \u001b[0;32m~/Documents/VIMSS/ontology/LLMs/enrichgpt-results/venv/lib/python3.10/site-packages/pandas/core/frame.py:8872\u001b[0m, in \u001b[0;36mDataFrame.groupby\u001b[0;34m(self, by, axis, level, as_index, sort, group_keys, observed, dropna)\u001b[0m\n\u001b[1;32m   8869\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m level \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m by \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   8870\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mYou have to supply one of \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mby\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m and \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlevel\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m-> 8872\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mDataFrameGroupBy\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   8873\u001b[0m \u001b[43m    \u001b[49m\u001b[43mobj\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m   8874\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkeys\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mby\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   8875\u001b[0m \u001b[43m    \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   8876\u001b[0m \u001b[43m    \u001b[49m\u001b[43mlevel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlevel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   8877\u001b[0m \u001b[43m    \u001b[49m\u001b[43mas_index\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mas_index\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   8878\u001b[0m \u001b[43m    \u001b[49m\u001b[43msort\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msort\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   8879\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgroup_keys\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgroup_keys\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   8880\u001b[0m \u001b[43m    \u001b[49m\u001b[43mobserved\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mobserved\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   8881\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdropna\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdropna\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   8882\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/VIMSS/ontology/LLMs/enrichgpt-results/venv/lib/python3.10/site-packages/pandas/core/groupby/groupby.py:1273\u001b[0m, in \u001b[0;36mGroupBy.__init__\u001b[0;34m(self, obj, keys, axis, level, grouper, exclusions, selection, as_index, sort, group_keys, observed, dropna)\u001b[0m\n\u001b[1;32m   1270\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdropna \u001b[38;5;241m=\u001b[39m dropna\n\u001b[1;32m   1272\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m grouper \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 1273\u001b[0m     grouper, exclusions, obj \u001b[38;5;241m=\u001b[39m \u001b[43mget_grouper\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1274\u001b[0m \u001b[43m        \u001b[49m\u001b[43mobj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1275\u001b[0m \u001b[43m        \u001b[49m\u001b[43mkeys\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1276\u001b[0m \u001b[43m        \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1277\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlevel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlevel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1278\u001b[0m \u001b[43m        \u001b[49m\u001b[43msort\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msort\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1279\u001b[0m \u001b[43m        \u001b[49m\u001b[43mobserved\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mobserved\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mis\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mlib\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mno_default\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mobserved\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1280\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdropna\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdropna\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1281\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1283\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m observed \u001b[38;5;129;01mis\u001b[39;00m lib\u001b[38;5;241m.\u001b[39mno_default:\n\u001b[1;32m   1284\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28many\u001b[39m(ping\u001b[38;5;241m.\u001b[39m_passed_categorical \u001b[38;5;28;01mfor\u001b[39;00m ping \u001b[38;5;129;01min\u001b[39;00m grouper\u001b[38;5;241m.\u001b[39mgroupings):\n",
      "File \u001b[0;32m~/Documents/VIMSS/ontology/LLMs/enrichgpt-results/venv/lib/python3.10/site-packages/pandas/core/groupby/grouper.py:1009\u001b[0m, in \u001b[0;36mget_grouper\u001b[0;34m(obj, key, axis, level, sort, observed, validate, dropna)\u001b[0m\n\u001b[1;32m   1007\u001b[0m         in_axis, level, gpr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m, gpr, \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1008\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1009\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(gpr)\n\u001b[1;32m   1010\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(gpr, Grouper) \u001b[38;5;129;01mand\u001b[39;00m gpr\u001b[38;5;241m.\u001b[39mkey \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1011\u001b[0m     \u001b[38;5;66;03m# Add key to exclusions\u001b[39;00m\n\u001b[1;32m   1012\u001b[0m     exclusions\u001b[38;5;241m.\u001b[39madd(gpr\u001b[38;5;241m.\u001b[39mkey)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'model'"
     ]
    }
   ],
   "source": [
    "subsets_df.fillna(0).groupby([MODEL, METHOD]).mean(numeric_only=True).style.highlight_max(axis=1, props='font-weight:bold').format(precision=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19c45026",
   "metadata": {},
   "outputs": [],
   "source": [
    "def agg_table(this_df, cols, exclude=[None]):\n",
    "    qcols = [MODEL, METHOD] + cols\n",
    "    agg_df = this_df.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD]).mean(numeric_only=True)\n",
    "    for x in exclude:\n",
    "        agg_df= agg_df.query(f\"method != '{x}'\")\n",
    "    return agg_df.reset_index().style.highlight_max(axis=0, props='font-weight:bold').format(precision=3)\n",
    "\n",
    "pd.options.display.precision = 2\n",
    "pd.set_option(\"display.precision\", 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d20be824",
   "metadata": {},
   "source": [
    "## TABLE: MAIN RESULTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c67dfd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "CORE_METRICS = [PROPORTION_SIGNIFICANT, HAS_TOP_HIT, NUM_GO_TERMS, NUM_NOVEL, NUM_UNPARSED]\n",
    "EXCLUDE = [\"standard\", \"standard no ontology\", \"random\", \"rank based\", \"closure\"]\n",
    "agg_table(df, CORE_METRICS, EXCLUDE)\n",
    "#means.query(\"method != 'standard'\").style.highlight_max(axis=0, props='font-weight:bold').format(precision=3)\n",
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "607496a4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "qcols = [MODEL, METHOD] + CORE_METRICS\n",
    "\n",
    "print(METHOD)\n",
    "#agg_df_orig = df.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD]).mean(numeric_only=True)\n",
    "#agg_df_plot = pd.DataFrame( columns=agg_df_orig.columns)#NULL#pd.DataFrame() \n",
    "#print(EXCLUDE)\n",
    "#agg_df_plot = agg_df_plot.query(f\"method in @METHOD\")\n",
    "\n",
    "\n",
    "print(qcols)\n",
    "agg_df_plot = df.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD]).mean(numeric_only=True)\n",
    "for x in EXCLUDE:\n",
    "    #print(x)\n",
    "    agg_df_plot= agg_df_plot.query(f\"method != '{x}'\")\n",
    "    \n",
    "print(agg_df_plot.shape)\n",
    "print(\"cols\")\n",
    "print(agg_df_plot.columns)\n",
    "\n",
    "print(\"index\")\n",
    "print(agg_df_plot.index)\n",
    "\n",
    "newindex = agg_df_plot.index\n",
    "#len = agg_df.index.to_flat_index().size()\n",
    "#print(len)\n",
    "#for i in range(1,len):\n",
    "#    newindex[i] = agg_df.index[i]\n",
    "\n",
    "#print(agg_df.loc[:, 'model'])\n",
    "\n",
    "#print(agg_df['num GO terms'])\n",
    "\n",
    "#rework for significant GO terms\n",
    "xs = agg_df_plot['num GO terms']/agg_df_plot['num unannotated']\n",
    "ys = agg_df_plot['num unparsed']/agg_df_plot['num unannotated']\n",
    "#print(agg_df.index)\n",
    "\n",
    "print(xs)\n",
    "yells = [\"yellow\"]*5\n",
    "reds = [\"red\"] *3\n",
    "purps = [\"purple\"] *3\n",
    "orans = [\"orange\"]*3\n",
    "#colors = [*yells, *reds, *purps, *orans]\n",
    "colors = [*reds, *purps, *orans]\n",
    "print(colors)\n",
    "\n",
    "plt.figure(figsize=(6,6))\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "ax.scatter(xs,ys,\n",
    "           c=colors)\n",
    "#, label=agg_df.index\n",
    "\n",
    "for i in range(0,len(xs)):\n",
    "    ax.annotate(agg_df_plot.index[i], (xs[i], ys[i]))\n",
    "    \n",
    "ax.set_xlabel(\"True terms/Error rate\")\n",
    "ax.set_ylabel(\"Unparsed terms/Error rate\")\n",
    "\n",
    "#handles, labels = ax.get_legend_handles_labels()\n",
    "#ax.legend(handles, labels, loc='upper left')\n",
    "\n",
    "plt.xlim(0,75)\n",
    "plt.ylim(0,75)\n",
    "\n",
    "plt.savefig(\"true_vs_unparsed_vs_error.pdf\", format=\"pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14b939c3",
   "metadata": {},
   "source": [
    "### Core subsets\n",
    "\n",
    "Copy this to [subsets tab](https://docs.google.com/spreadsheets/d/1gGO5IHEg-N0hivtHBO6-rdXtin8hPhw-zv6eYOBgXcE/edit#gid=669935942)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4455f40e",
   "metadata": {},
   "outputs": [],
   "source": [
    "subsets_cols =  [MODEL, METHOD, \"goslim_generic\", \"anc_of_goslim_generic\", \"goslim_agr\", \"anc_of_goslim_agr\"]\n",
    "subsets_grouped = subsets_df.fillna(0).groupby([MODEL, METHOD])[subsets_cols]\n",
    "means = subsets_grouped.mean(numeric_only=True)\n",
    "means[\"closure_of_goslim_generic\"] = means[\"goslim_generic\"] + means[\"anc_of_goslim_generic\"]\n",
    "means[\"closure_of_goslim_agr\"] = means[\"goslim_agr\"] + means[\"anc_of_goslim_agr\"]\n",
    "means.style.highlight_max(axis=0, props='font-weight:bold').format(precision=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdf29a42",
   "metadata": {},
   "source": [
    "### TABLE: Above the shoreline in key subsets\n",
    "\n",
    "[anc subsets tab](https://docs.google.com/spreadsheets/d/1gGO5IHEg-N0hivtHBO6-rdXtin8hPhw-zv6eYOBgXcE/edit#gid=345667144)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a0e3b53",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sorted_means = means.sort_values(\"anc_of_goslim_generic\", ascending=False)\n",
    "filtered = sorted_means[[\"anc_of_goslim_generic\", \"anc_of_goslim_agr\"]]\n",
    "filtered.style.highlight_max(axis=0, props='font-weight:bold').format(precision=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b527e9ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "means[[\"anc_of_goslim_generic\", \"anc_of_goslim_agr\"]].sort_values(\"anc_of_goslim_agr\", ascending=True).style.highlight_min(axis=0, props='font-weight:bold').format(precision=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45375418",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reset the index of the pivoted DataFrame to have 'profession' as a column\n",
    "#subsets_df = subsets_df.fillna(0).groupby([MODEL, METHOD])[subsets_cols].mean(numeric_only=True).reset_index()\n",
    "\n",
    "# Melt the DataFrame to have 'category', 'profession', and 'percentage' columns\n",
    "#melted_df = subsets_df.melt(id_vars=[MODEL, METHOD], var_name='subset', value_name='proportion')\n",
    "#melted_df[\"mm\"] = melted_df[MODEL] + melted_df[METHOD]\n",
    "#melted_df\n",
    "# Create a bar plot using Seaborn\n",
    "#plt.figure(figsize=(10, 6))\n",
    "#sns.barplot(x='subset', y='proportion', hue=\"mm\", data=melted_df)\n",
    "#plt.title('Subsets by method')\n",
    "#plt.xlabel('Subset')\n",
    "#plt.ylabel('Proportion')\n",
    "#plt.legend(title='Subsets')\n",
    "#plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ba3d6b3",
   "metadata": {},
   "source": [
    "## Evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38e69d5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "agg_table(df, [NUM_GO_TERMS, SIZE_OVERLAP, NR_SIZE_OVERLAP], [\"standard\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e4a701c",
   "metadata": {},
   "outputs": [],
   "source": [
    "eval_summary_cols = [HAS_TOP_HIT, IN_TOP_5, IN_TOP_10, SIZE_OVERLAP, SIMILARITY, NUM_TERMS, NUM_GO_TERMS, NR_SIZE_OVERLAP, NR_SIMILARITY, MEAN_P_VALUE, MIN_P_VALUE, MAX_P_VALUE, PROPORTION_SIGNIFICANT, NUM_NOVEL]\n",
    "agg_table(df, eval_summary_cols)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc1fec78",
   "metadata": {},
   "source": [
    "## as above, no perturbation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9993e1ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_no_perturb = df.query(f\"{GENE_RANDOMIZATION_FACTOR} == 0\")\n",
    "agg_table(df_no_perturb, CORE_METRICS, EXCLUDE)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc3bdb22",
   "metadata": {},
   "source": [
    "## Maximums"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f251102",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[[MODEL, METHOD] + eval_summary_cols].groupby([MODEL, METHOD]).max(numeric_only=True).style.highlight_max(axis=0, props='font-weight:bold').format(precision=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88ef67be",
   "metadata": {},
   "source": [
    "### Effect of truncation\n",
    "\n",
    "Larger gene sets penalize annotation-based GPT methods due to the necessity to truncate to fit in the window size.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "639bcfc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sns.set(color_codes=True)\n",
    "np.random.seed(sum(map(ord, \"regression\")))\n",
    "sns.lmplot(x=GENESET_SIZE, y=TRUNCATION_FACTOR, data=df.query(\"method=='ontological_synopsis'\"))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "165cbace",
   "metadata": {},
   "outputs": [],
   "source": [
    "dftups = [(method, df.query(f\"method=='{method}'\")) for method in [\"ontological_synopsis\", \"narrative_synopsis\"]]\n",
    "for m, mdf in dftups:\n",
    "    mdf.assign(dataset=m)\n",
    "concatenated = pd.concat([mdf.assign(dataset=m) for m, mdf in dftups])\n",
    "sns.scatterplot(x=GENESET_SIZE, y=TRUNCATION_FACTOR, data=concatenated, style='dataset', hue='model')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33fdd5fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.query(f\"{GENESET_SIZE} < 50\")[[SOURCE_GENESET]].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04e7e4f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_small = df.query(f\"{GENESET_SIZE} < 50\")\n",
    "means = df_small[[MODEL, METHOD] + eval_summary_cols].groupby(['model', 'method']).mean(numeric_only=True)\n",
    "means"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70db39e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def color_lightness(rgb, lightness_scale):\n",
    "    h, l, s = colorsys.rgb_to_hls(*rgb)\n",
    "    return colorsys.hls_to_rgb(h, min(1, l * lightness_scale), s = s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43d877f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "agg_df_plot = df.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD]).mean(numeric_only=True)\n",
    "for x in EXCLUDE:\n",
    "    #print(x)\n",
    "    agg_df_plot = agg_df_plot.query(f\"method != '{x}'\")    \n",
    "    \n",
    "agg_df_small_plot = df_small.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD]).mean(numeric_only=True)\n",
    "for x in EXCLUDE:\n",
    "    #print(x)\n",
    "    agg_df_small_plot = agg_df_small_plot.query(f\"method != '{x}'\")\n",
    "    \n",
    "agg_df_plot_std = df.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD]).std(numeric_only=True)\n",
    "for x in EXCLUDE:\n",
    "    #print(x)\n",
    "    agg_df_plot_std = agg_df_plot_std.query(f\"method != '{x}'\")    \n",
    "    \n",
    "agg_df_small_plot_std = df_small.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD]).std(numeric_only=True)\n",
    "for x in EXCLUDE:\n",
    "    #print(x)\n",
    "    agg_df_small_plot_std = agg_df_small_plot_std.query(f\"method != '{x}'\")\n",
    "\n",
    "\n",
    "print(agg_df_small_plot.shape)\n",
    "print(\"cols\")\n",
    "print(agg_df_small_plot.columns)\n",
    "\n",
    "print(\"index\")\n",
    "print(agg_df_small_plot.index)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "635a65df",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "qcols = [MODEL, METHOD] + CORE_METRICS\n",
    "\n",
    "print(METHOD)\n",
    "#agg_df_orig = df.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD]).mean(numeric_only=True)\n",
    "#agg_df = pd.DataFrame( columns=agg_df_orig.columns)#NULL#pd.DataFrame() \n",
    "#print(EXCLUDE)\n",
    "#agg_df = agg_df_orig.query(f\"method in @METHOD\")\n",
    "\n",
    "print(qcols)\n",
    "\n",
    "newindex = agg_df_plot.index\n",
    "#len = agg_df.index.to_flat_index().size()\n",
    "#print(len)\n",
    "#for i in range(1,len):\n",
    "#    newindex[i] = agg_df.index[i]\n",
    "\n",
    "#print(agg_df.loc[:, 'model'])\n",
    "\n",
    "#print(agg_df['num GO terms'])\n",
    "\n",
    "#rework for significant GO terms\n",
    "xs = agg_df_plot['num GO terms']/agg_df_plot['num unannotated']\n",
    "ys = agg_df_plot['num unparsed']/agg_df_plot['num unannotated']\n",
    "\n",
    "xs_small = agg_df_small_plot['num GO terms']/agg_df_small_plot['num unannotated']\n",
    "ys_small = agg_df_small_plot['num unparsed']/agg_df_small_plot['num unannotated']\n",
    "\n",
    "#print(agg_df.index)\n",
    "\n",
    "print(xs)\n",
    "yells = [\"yellow\"]*5\n",
    "reds = [\"red\"] *3\n",
    "purps = [\"purple\"] *3\n",
    "orans = [\"orange\"]*3\n",
    "\n",
    "colorr = mpl.colors.ColorConverter.to_rgb(\"red\")\n",
    "rgbr = [color_lightness(colorr, scale) for scale in [2]]\n",
    "colorp = mpl.colors.ColorConverter.to_rgb(\"purple\")\n",
    "rgbp = [color_lightness(colorp, scale) for scale in [2]]\n",
    "coloro = mpl.colors.ColorConverter.to_rgb(\"orange\")\n",
    "rgbo = [color_lightness(coloro, scale) for scale in [2]]\n",
    "\n",
    "redss = rgbr *3\n",
    "purpss = rgbp *3\n",
    "oranss = rgbo *3\n",
    "\n",
    "\n",
    "#colors = [*yells, *reds, *purps, *orans]\n",
    "colors = [*reds, *purps, *orans]\n",
    "print(colors)\n",
    "\n",
    "colors_small = [*redss, *purpss, *oranss]\n",
    "print(colors_small)\n",
    "\n",
    "circles= ['o','^','s']*3\n",
    "#triangles= ['^']*3\n",
    "#squares= ['s']*3\n",
    "\n",
    "markers = [*circles]#, *triangles, *squares]\n",
    "\n",
    "SMALL_SIZE = 6\n",
    "plt.rc('font', size=SMALL_SIZE)   \n",
    "plt.figure(figsize=(6,6))\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "#ax.scatter(xs,ys, c=colors, s=50, marker=marker)\n",
    "#ax.scatter(xs_small,xs_small, c=colors, s=25, marker=marker)\n",
    "\n",
    "unique_markers = set(markers)\n",
    "print(unique_markers)\n",
    "print(markers)\n",
    "\n",
    "for um in unique_markers:\n",
    "    #print(um)\n",
    "    #print(markers.index(um))    \n",
    "    #print([index for index, elem in enumerate(markers) if elem == um])\n",
    "    mask = [index for index, elem in enumerate(markers) if elem == um]# markers == um\n",
    "    maskar = np.array(mask)\n",
    "    print(mask)\n",
    "    print(type(mask))\n",
    "    print(maskar)\n",
    "    print(type(maskar))\n",
    "    minv = mask[0]\n",
    "    maxv = mask[2]\n",
    "    print(str(minv)+\"\\t\"+str(maxv))\n",
    "    nowcol = [colors[mask[0]], colors[mask[1]], colors[mask[2]]]\n",
    "    print(nowcol)\n",
    "    \n",
    "    ax.scatter(xs[mask], ys[mask], marker=um, c=nowcol, s=150, alpha=0.5, edgecolors=\"black\",  linewidth=0.5)\n",
    "    ax.scatter(xs_small[mask], ys_small[mask], marker=um, c=nowcol, s=50, alpha=0.5, edgecolors=\"black\",  linewidth=0.5)\n",
    "    \n",
    "#, label=agg_df.index\n",
    "agg_df_plot_labels = agg_df_plot.index.values\n",
    "for i in range(0, len(xs)):\n",
    "    agg_df_plot_labels[i] = (agg_df_plot_labels[i][0] + \": \" + agg_df_plot_labels[i][1]).replace(\"(\",\"\")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\")\",\"\")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\" - \",\" : \")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\"gpt\",\"GPT\")        \n",
    "                \n",
    "for i in range(0,len(xs)):\n",
    "    ax.annotate(agg_df_plot_labels[i], (xs[i], ys[i]))\n",
    "    ax.annotate(agg_df_plot_labels[i], (xs_small[i], ys_small[i]))  \n",
    "ax.set_xlabel(\"True terms/Error\")\n",
    "ax.set_ylabel(\"Unparsed terms/Error\")\n",
    "\n",
    "#handles, labels = ax.get_legend_handles_labels()\n",
    "#ax.legend(handles, labels, loc='upper left')\n",
    "\n",
    "plt.xlim(5,70)\n",
    "plt.ylim(5,70)\n",
    "plt.gca().set_aspect(\"equal\")\n",
    "\n",
    "plt.savefig(\"true_vs_unparsed_vs_error.pdf\", format=\"pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77540ff4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "qcols = [MODEL, METHOD] + CORE_METRICS\n",
    "\n",
    "print(METHOD)\n",
    "#agg_df_orig = df.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD]).mean(numeric_only=True)\n",
    "#agg_df = pd.DataFrame( columns=agg_df_orig.columns)#NULL#pd.DataFrame() \n",
    "#print(EXCLUDE)\n",
    "#agg_df = agg_df_orig.query(f\"method in @METHOD\")\n",
    "\n",
    "print(qcols)\n",
    "\n",
    "newindex = agg_df_plot.index\n",
    "#len = agg_df.index.to_flat_index().size()\n",
    "#print(len)\n",
    "#for i in range(1,len):\n",
    "#    newindex[i] = agg_df.index[i]\n",
    "\n",
    "#print(agg_df.loc[:, 'model'])\n",
    "\n",
    "#print(agg_df['num GO terms'])\n",
    "\n",
    "#rework for significant GO terms\n",
    "xs = agg_df_plot['has top term']/agg_df_plot['num unannotated']\n",
    "ys = agg_df_plot['proportion significant']/agg_df_plot['num unannotated']\n",
    "\n",
    "xs_small = agg_df_small_plot['has top term']/agg_df_small_plot['num unannotated']\n",
    "ys_small = agg_df_small_plot['proportion significant']/agg_df_small_plot['num unannotated']\n",
    "\n",
    "#print(agg_df.index)\n",
    "\n",
    "print(xs)\n",
    "print(ys)\n",
    "print(np.max(xs))\n",
    "print(np.max(ys))\n",
    "print(np.max(xs_small))\n",
    "print(np.max(ys_small))\n",
    "\n",
    "\n",
    "yells = [\"yellow\"]*5\n",
    "reds = [\"red\"] *3\n",
    "purps = [\"purple\"] *3\n",
    "orans = [\"orange\"]*3\n",
    "\n",
    "colorr = mpl.colors.ColorConverter.to_rgb(\"red\")\n",
    "rgbr = [color_lightness(colorr, scale) for scale in [2]]\n",
    "colorp = mpl.colors.ColorConverter.to_rgb(\"purple\")\n",
    "rgbp = [color_lightness(colorp, scale) for scale in [2]]\n",
    "coloro = mpl.colors.ColorConverter.to_rgb(\"orange\")\n",
    "rgbo = [color_lightness(coloro, scale) for scale in [2]]\n",
    "\n",
    "redss = rgbr *3\n",
    "purpss = rgbp *3\n",
    "oranss = rgbo *3\n",
    "\n",
    "\n",
    "#colors = [*yells, *reds, *purps, *orans]\n",
    "colors = [*reds, *purps, *orans]\n",
    "print(colors)\n",
    "\n",
    "colors_small = [*redss, *purpss, *oranss]\n",
    "print(colors_small)\n",
    "\n",
    "circles= ['o','^','s']*3\n",
    "#triangles= ['^']*3\n",
    "#squares= ['s']*3\n",
    "\n",
    "markers = [*circles]#, *triangles, *squares]\n",
    "\n",
    "SMALL_SIZE = 6\n",
    "plt.rc('font', size=SMALL_SIZE)   \n",
    "plt.figure(figsize=(6,6))\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "#ax.scatter(xs,ys, c=colors, s=50, marker=marker)\n",
    "#ax.scatter(xs_small,xs_small, c=colors, s=25, marker=marker)\n",
    "\n",
    "unique_markers = set(markers)\n",
    "print(unique_markers)\n",
    "print(markers)\n",
    "\n",
    "for um in unique_markers:\n",
    "    #print(um)\n",
    "    #print(markers.index(um))    \n",
    "    #print([index for index, elem in enumerate(markers) if elem == um])\n",
    "    mask = [index for index, elem in enumerate(markers) if elem == um]# markers == um\n",
    "    maskar = np.array(mask)\n",
    "    print(mask)\n",
    "    print(type(mask))\n",
    "    print(maskar)\n",
    "    print(type(maskar))\n",
    "    minv = mask[0]\n",
    "    maxv = mask[2]\n",
    "    print(str(minv)+\"\\t\"+str(maxv))\n",
    "    nowcol = [colors[mask[0]], colors[mask[1]], colors[mask[2]]]\n",
    "    print(nowcol)\n",
    "    \n",
    "    ax.scatter(xs[mask], ys[mask], marker=um, c=nowcol, s=150, alpha=0.5, edgecolors=\"black\",  linewidth=0.5)\n",
    "    ax.scatter(xs_small[mask], ys_small[mask], marker=um, c=nowcol, s=50, alpha=0.5, edgecolors=\"black\",  linewidth=0.5)\n",
    "    \n",
    "#, label=agg_df.index\n",
    "agg_df_plot_labels = agg_df_plot.index.values\n",
    "for i in range(0, len(xs)):\n",
    "    agg_df_plot_labels[i] = (agg_df_plot_labels[i][0] + \": \" + agg_df_plot_labels[i][1]).replace(\"(\",\"\")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\")\",\"\")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\" - \",\" : \")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\"gpt\",\"GPT\")        \n",
    "                \n",
    "for i in range(0,len(xs)):\n",
    "    ax.annotate(agg_df_plot_labels[i], (xs[i], ys[i]))\n",
    "    ax.annotate(agg_df_plot_labels[i], (xs_small[i], ys_small[i]))  \n",
    "ax.set_xlabel(\"Has top term/Error\")\n",
    "ax.set_ylabel(\"Proportion significant/Error\")\n",
    "\n",
    "#handles, labels = ax.get_legend_handles_labels()\n",
    "#ax.legend(handles, labels, loc='upper left')\n",
    "\n",
    "plt.xlim(0,7)\n",
    "plt.ylim(0,7)\n",
    "plt.gca().set_aspect(\"equal\")\n",
    "\n",
    "plt.savefig(\"top_vs_proportion_vs_error.pdf\", format=\"pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f13b2a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "qcols = [MODEL, METHOD] + CORE_METRICS\n",
    "\n",
    "print(METHOD)\n",
    "#agg_df_orig = df.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD]).mean(numeric_only=True)\n",
    "#agg_df = pd.DataFrame( columns=agg_df_orig.columns)#NULL#pd.DataFrame() \n",
    "#print(EXCLUDE)\n",
    "#agg_df = agg_df_orig.query(f\"method in @METHOD\")\n",
    "\n",
    "print(qcols)\n",
    "\n",
    "agg_df_plot_all = df.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD])\n",
    "\n",
    "#pd.concat(map(lambda x: x[1], groups)).sort_index(\n",
    "agg_df_plot_all = agg_df_plot_all.apply(lambda x: x)\n",
    "#\n",
    "#agg_df_plot_all = agg_df_plot_all.head(1000000)\n",
    "\n",
    "print(agg_df_plot_all.shape)\n",
    "type(agg_df_plot_all)\n",
    "#print(agg_df_plot_all.iloc[:200])\n",
    "for x in EXCLUDE:\n",
    "    #print(x)\n",
    "    agg_df_plot_all = agg_df_plot_all.query(f\"method != '{x}'\")    \n",
    "    \n",
    "agg_df_small_plot_all = df_small.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD])\n",
    "agg_df_small_plot_all = agg_df_small_plot_all.apply(lambda x: x)\n",
    "\n",
    "for x in EXCLUDE:\n",
    "    #print(x)\n",
    "    agg_df_small_plot_all = agg_df_small_plot_all.query(f\"method != '{x}'\")\n",
    "    \n",
    "newindex = agg_df_plot_all.index\n",
    "#len = agg_df.index.to_flat_index().size()\n",
    "#print(len)\n",
    "#for i in range(1,len):\n",
    "#    newindex[i] = agg_df.index[i]\n",
    "\n",
    "#print(agg_df.loc[:, 'model'])\n",
    "#print(agg_df['num GO terms'])\n",
    "\n",
    "#rework for significant GO terms\n",
    "xs = agg_df_plot_all['has top term']\n",
    "ys = agg_df_plot_all['proportion significant']\n",
    "\n",
    "xs_small = agg_df_small_plot_all['has top term']\n",
    "ys_small = agg_df_small_plot_all['proportion significant']\n",
    "\n",
    "#print(agg_df.index)\n",
    "\n",
    "#print(xs)\n",
    "#print(ys)\n",
    "#print(np.max(xs))\n",
    "#print(np.max(ys))\n",
    "#print(np.max(xs_small))\n",
    "#print(np.max(ys_small))\n",
    "\n",
    "\n",
    "#yells = [\"yellow\"]*5\n",
    "reds = [\"red\"] *3\n",
    "purps = [\"purple\"] *3\n",
    "orans = [\"orange\"]*3\n",
    "\n",
    "colorr = mpl.colors.ColorConverter.to_rgb(\"red\")\n",
    "rgbr = [color_lightness(colorr, scale) for scale in [2]]\n",
    "colorp = mpl.colors.ColorConverter.to_rgb(\"purple\")\n",
    "rgbp = [color_lightness(colorp, scale) for scale in [2]]\n",
    "coloro = mpl.colors.ColorConverter.to_rgb(\"orange\")\n",
    "rgbo = [color_lightness(coloro, scale) for scale in [2]]\n",
    "\n",
    "redss = rgbr *3\n",
    "purpss = rgbp *3\n",
    "oranss = rgbo *3\n",
    "\n",
    "\n",
    "#colors = [*yells, *reds, *purps, *orans]\n",
    "colors = [*reds, *purps, *orans]\n",
    "print(colors)\n",
    "\n",
    "colors_small = [*redss, *purpss, *oranss]\n",
    "print(colors_small)\n",
    "\n",
    "circles= ['o','^','s']*3\n",
    "#triangles= ['^']*3\n",
    "#squares= ['s']*3\n",
    "\n",
    "markers = [*circles]#, *triangles, *squares]\n",
    "\n",
    "SMALL_SIZE = 6\n",
    "plt.rc('font', size=SMALL_SIZE)   \n",
    "plt.figure(figsize=(6,6))\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "#ax.scatter(xs,ys, c=colors, s=50, marker=marker)\n",
    "#ax.scatter(xs_small,xs_small, c=colors, s=25, marker=marker)\n",
    "\n",
    "unique_markers = set(markers)\n",
    "print(unique_markers)\n",
    "print(markers)\n",
    "\n",
    "for um in unique_markers:\n",
    "    #print(um)\n",
    "    #print(markers.index(um))    \n",
    "    #print([index for index, elem in enumerate(markers) if elem == um])\n",
    "    mask = [index for index, elem in enumerate(markers) if elem == um]# markers == um\n",
    "    maskar = np.array(mask)\n",
    "    #print(mask)\n",
    "    #print(type(mask))\n",
    "    #print(maskar)\n",
    "    #print(type(maskar))\n",
    "    minv = mask[0]\n",
    "    maxv = mask[2]\n",
    "    #print(str(minv)+\"\\t\"+str(maxv))\n",
    "    nowcol = [colors[mask[0]], colors[mask[1]], colors[mask[2]]]\n",
    "    #print(nowcol)\n",
    "    \n",
    "    ax.scatter(xs[mask], ys[mask], marker=um, c=nowcol, s=150, alpha=0.5, edgecolors=\"black\",  linewidth=0.5)\n",
    "    ax.scatter(xs_small[mask], ys_small[mask], marker=um, c=nowcol, s=50, alpha=0.5, edgecolors=\"black\",  linewidth=0.5)\n",
    "    \n",
    "#, label=agg_df.index\n",
    "agg_df_plot_labels = agg_df_plot_all.index.values\n",
    "for i in range(0, len(xs)):\n",
    "    agg_df_plot_labels[i] = (agg_df_plot_labels[i][0] + \": \" + agg_df_plot_labels[i][1]).replace(\"(\",\"\")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\")\",\"\")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\" - \",\" : \")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\"gpt\",\"GPT\")        \n",
    "                \n",
    "for i in range(0,len(xs)):\n",
    "    ax.annotate(agg_df_plot_labels[i], (xs[i], ys[i]))\n",
    "for i in range(0,len(xs_small)):\n",
    "    ax.annotate(agg_df_plot_labels[i], (xs_small[i], ys_small[i]))  \n",
    "\n",
    "ax.set_xlabel(\"Has top term/Error\")\n",
    "ax.set_ylabel(\"Proportion significant/Error\")\n",
    "\n",
    "#handles, labels = ax.get_legend_handles_labels()\n",
    "#ax.legend(handles, labels, loc='upper left')\n",
    "\n",
    "plt.xlim(0,1)\n",
    "plt.ylim(0,1)\n",
    "plt.gca().set_aspect(\"equal\")\n",
    "\n",
    "plt.savefig(\"top_vs_proportion__individual.pdf\", format=\"pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4f52ee6-c882-4935-90ee-69480340626b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns\n",
    "#num GO terms = TP\n",
    "#num unannotated = FP\n",
    "#num terms\n",
    "#num unparsed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80ab1eee-8479-4d89-ac41-251058e8f03d",
   "metadata": {},
   "outputs": [],
   "source": [
    "CORE_METRICS2 = [FP, TP]\n",
    "qcols = [MODEL, METHOD] + CORE_METRICS2\n",
    "\n",
    "print(METHOD)\n",
    "print(qcols)\n",
    "\n",
    "agg_df_plot_tpfp = df.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD]).mean(numeric_only=True)\n",
    "\n",
    "agg_df_plot_tpfp = agg_df_plot_tpfp.apply(lambda x: x)\n",
    "\n",
    "print(agg_df_plot_tpfp.shape)\n",
    "type(agg_df_plot_tpfp)\n",
    "for x in EXCLUDE:\n",
    "    agg_df_plot_tpfp = agg_df_plot_tpfp.query(f\"method != '{x}'\")    \n",
    "\n",
    "\n",
    "agg_df_small_plot_tpfp = df_small.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD]).mean(numeric_only=True)\n",
    "agg_df_small_plot_tpfp = agg_df_small_plot_tpfp.apply(lambda x: x)\n",
    "print(agg_df_small_plot_tpfp.columns)\n",
    "\n",
    "for x in EXCLUDE:\n",
    "    agg_df_small_plot_tpfp = agg_df_small_plot_tpfp.query(f\"method != '{x}'\")\n",
    "    \n",
    "newindex = agg_df_plot_tpfp.index\n",
    "\n",
    "xs = agg_df_plot_tpfp['true positive rate across a set of gene GO annotations']\n",
    "ys = agg_df_plot_tpfp['false positive rate across a set of gene GO annotations']\n",
    "\n",
    "xs_small = agg_df_small_plot_tpfp['true positive rate across a set of gene GO annotations']\n",
    "ys_small = agg_df_small_plot_tpfp['false positive rate across a set of gene GO annotations']\n",
    "\n",
    "reds = [\"red\"] *3\n",
    "purps = [\"purple\"] *3\n",
    "orans = [\"orange\"]*3\n",
    "\n",
    "colorr = mpl.colors.ColorConverter.to_rgb(\"red\")\n",
    "rgbr = [color_lightness(colorr, scale) for scale in [2]]\n",
    "colorp = mpl.colors.ColorConverter.to_rgb(\"purple\")\n",
    "rgbp = [color_lightness(colorp, scale) for scale in [2]]\n",
    "coloro = mpl.colors.ColorConverter.to_rgb(\"orange\")\n",
    "rgbo = [color_lightness(coloro, scale) for scale in [2]]\n",
    "\n",
    "redss = rgbr *3\n",
    "purpss = rgbp *3\n",
    "oranss = rgbo *3\n",
    "\n",
    "colors = [*reds, *purps, *orans]\n",
    "print(colors)\n",
    "\n",
    "colors_small = [*redss, *purpss, *oranss]\n",
    "print(colors_small)\n",
    "\n",
    "circles= ['o','^','s']*3\n",
    "\n",
    "markers = [*circles]\n",
    "\n",
    "SMALL_SIZE = 6\n",
    "plt.figure()\n",
    "plt.rc('font', size=SMALL_SIZE)   \n",
    "plt.figure(figsize=(10,6))\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "unique_markers = set(markers)\n",
    "print(unique_markers)\n",
    "print(markers)\n",
    "\n",
    "for um in unique_markers:\n",
    "    mask = [index for index, elem in enumerate(markers) if elem == um]\n",
    "    maskar = np.array(mask)\n",
    "    minv = mask[0]\n",
    "    maxv = mask[2]\n",
    "    nowcol = [colors[mask[0]], colors[mask[1]], colors[mask[2]]]\n",
    "    \n",
    "    ax.scatter(xs[mask], ys[mask], marker=um, c=nowcol, s=150, alpha=0.5, edgecolors=\"black\",  linewidth=0.5)#, marker=um#, c=nowcol\n",
    "    ax.scatter(xs_small[mask], ys_small[mask], marker=um, c=nowcol, s=50, alpha=0.5, edgecolors=\"black\",  linewidth=0.5)#, marker=um#, c=nowcol\n",
    "    \n",
    "agg_df_plot_labels = agg_df_plot_tpfp.index.values\n",
    "for i in range(0, len(xs)):\n",
    "    agg_df_plot_labels[i] = (agg_df_plot_labels[i][0] + \": \" + agg_df_plot_labels[i][1]).replace(\"(\",\"\")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\")\",\"\")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\" - \",\" : \")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\"gpt\",\"GPT\")        \n",
    "                \n",
    "for i in range(0,len(xs)):\n",
    "    ax.annotate(agg_df_plot_labels[i], (xs[i], ys[i]))\n",
    "for i in range(0,len(xs_small)):\n",
    "    ax.annotate(agg_df_plot_labels[i], (xs_small[i], ys_small[i]))  \n",
    "\n",
    "xmax = 0.3\n",
    "ymax = 0.6\n",
    "# Major ticks every 20, minor ticks every 5\n",
    "xmajor_ticks = np.arange(0, xmax, 0.05)\n",
    "xminor_ticks = np.arange(0, xmax, 0.01)\n",
    "ymajor_ticks = np.arange(ymax, 1, 0.1)\n",
    "yminor_ticks = np.arange(ymax, 1, 0.05)\n",
    "\n",
    "ax.set_xticks(xmajor_ticks)\n",
    "ax.set_xticks(xminor_ticks, minor=True)\n",
    "ax.set_yticks(ymajor_ticks)\n",
    "ax.set_yticks(yminor_ticks, minor=True)\n",
    "\n",
    "# And a corresponding grid\n",
    "ax.grid(which='both')\n",
    "\n",
    "# Or if you want different settings for the grids:\n",
    "ax.grid(which='minor', alpha=0.5)\n",
    "ax.grid(which='major', alpha=0.5)\n",
    "#ax.grid(color=\"blue\")\n",
    "\n",
    "\n",
    "ax.set_xlabel(\"FP\")\n",
    "ax.set_ylabel(\"TP\")\n",
    "\n",
    "plt.xlim(0,xmax)\n",
    "plt.ylim(ymax,1)\n",
    "#plt.gca().set_aspect(\"equal\")\n",
    "\n",
    "plt.savefig(\"TP_vs_FP.pdf\", format=\"pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d0977c5-178c-447e-a90f-b414b9674609",
   "metadata": {},
   "outputs": [],
   "source": [
    "CORE_METRICS3 = [RECALL, PRECISION]\n",
    "qcols = [MODEL, METHOD] + CORE_METRICS3\n",
    "\n",
    "print(METHOD)\n",
    "print(qcols)\n",
    "\n",
    "agg_df_plot_pr = df.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD]).mean(numeric_only=True)\n",
    "agg_df_plot_pr = agg_df_plot_pr.apply(lambda x: x)\n",
    "\n",
    "print(agg_df_plot_pr.shape)\n",
    "type(agg_df_plot_pr)\n",
    "for x in EXCLUDE:\n",
    "    agg_df_plot_pr = agg_df_plot_pr.query(f\"method != '{x}'\")    \n",
    "\n",
    "print(agg_df_plot_pr.columns)\n",
    "\n",
    "agg_df_small_plot_pr = df_small.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD]).mean(numeric_only=True)\n",
    "agg_df_small_plot_pr = agg_df_small_plot_pr.apply(lambda x: x)\n",
    "\n",
    "for x in EXCLUDE:\n",
    "    agg_df_small_plot_pr = agg_df_small_plot_pr.query(f\"method != '{x}'\")\n",
    "    \n",
    "newindex = agg_df_plot_pr.index\n",
    "\n",
    "xs = agg_df_plot_pr['recall']\n",
    "ys = agg_df_plot_pr['precision']\n",
    "\n",
    "xs_small = agg_df_small_plot_pr['recall']\n",
    "ys_small = agg_df_small_plot_pr['precision']\n",
    "\n",
    "reds = [\"red\"] *3\n",
    "purps = [\"purple\"] *3\n",
    "orans = [\"orange\"]*3\n",
    "\n",
    "colorr = mpl.colors.ColorConverter.to_rgb(\"red\")\n",
    "rgbr = [color_lightness(colorr, scale) for scale in [2]]\n",
    "colorp = mpl.colors.ColorConverter.to_rgb(\"purple\")\n",
    "rgbp = [color_lightness(colorp, scale) for scale in [2]]\n",
    "coloro = mpl.colors.ColorConverter.to_rgb(\"orange\")\n",
    "rgbo = [color_lightness(coloro, scale) for scale in [2]]\n",
    "\n",
    "redss = rgbr *3\n",
    "purpss = rgbp *3\n",
    "oranss = rgbo *3\n",
    "\n",
    "colors = [*reds, *purps, *orans]\n",
    "print(colors)\n",
    "\n",
    "colors_small = [*redss, *purpss, *oranss]\n",
    "print(colors_small)\n",
    "\n",
    "circles= ['o','^','s']*3\n",
    "\n",
    "markers = [*circles]\n",
    "\n",
    "SMALL_SIZE = 6\n",
    "plt.figure()\n",
    "plt.rc('font', size=SMALL_SIZE)   \n",
    "plt.figure(figsize=(10,6))\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "unique_markers = set(markers)\n",
    "print(unique_markers)\n",
    "print(markers)\n",
    "\n",
    "for um in unique_markers:\n",
    "    mask = [index for index, elem in enumerate(markers) if elem == um]\n",
    "    maskar = np.array(mask)\n",
    "    minv = mask[0]\n",
    "    maxv = mask[2]\n",
    "    nowcol = [colors[mask[0]], colors[mask[1]], colors[mask[2]]]\n",
    "    \n",
    "    ax.scatter(xs[mask], ys[mask], marker=um, c=nowcol, s=200, alpha=0.5, edgecolors=\"black\",  linewidth=0.5)#, marker=um#, c=nowcol\n",
    "    ax.scatter(xs_small[mask], ys_small[mask], marker=um, c=\"white\", s=100, alpha=0.5, edgecolors=nowcol,  linewidth=2)#, marker=um#, c=nowcol\n",
    "    \n",
    "agg_df_plot_labels = agg_df_plot_pr.index.values\n",
    "for i in range(0, len(xs)):\n",
    "    agg_df_plot_labels[i] = (agg_df_plot_labels[i][0] + \": \" + agg_df_plot_labels[i][1]).replace(\"(\",\"\")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\")\",\"\")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\" - \",\" : \")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\"gpt\",\"GPT\")        \n",
    "                \n",
    "for i in range(0,len(xs)):\n",
    "    ax.annotate(agg_df_plot_labels[i], (xs[i], ys[i]))\n",
    "for i in range(0,len(xs_small)):\n",
    "    ax.annotate(agg_df_plot_labels[i], (xs_small[i], ys_small[i]))  \n",
    "\n",
    "xmax = 0.3\n",
    "ymax = 0.6\n",
    "# Major ticks every 20, minor ticks every 5\n",
    "xmajor_ticks = np.arange(0, 1, 0.05)\n",
    "xminor_ticks = np.arange(0, 1, 0.01)\n",
    "ymajor_ticks = np.arange(0, 1, 0.05)\n",
    "yminor_ticks = np.arange(0, 1, 0.01)\n",
    "\n",
    "ax.set_xticks(xmajor_ticks)\n",
    "ax.set_xticks(xminor_ticks, minor=True)\n",
    "ax.set_yticks(ymajor_ticks)\n",
    "ax.set_yticks(yminor_ticks, minor=True)\n",
    "\n",
    "# And a corresponding grid\n",
    "ax.grid(which='both')\n",
    "\n",
    "# Or if you want different settings for the grids:\n",
    "ax.grid(which='minor', alpha=0.5)\n",
    "ax.grid(which='major', alpha=0.5)\n",
    "#ax.grid(color=\"blue\")\n",
    "\n",
    "\n",
    "ax.set_xlabel(\"Recall\")\n",
    "ax.set_ylabel(\"Precision\")\n",
    "\n",
    "plt.xlim(0,0.1)\n",
    "plt.ylim(0.8,1)\n",
    "#plt.gca().set_aspect(\"equal\")\n",
    "\n",
    "plt.savefig(\"precision_vs_recall.pdf\", format=\"pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee33c5e7-e396-4383-8720-2df76a48f2b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "agg_df_plot_pr.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6d4b75d-4bfa-4cca-ab11-ad4427c73737",
   "metadata": {},
   "outputs": [],
   "source": [
    "CORE_METRICS3 = [RECALL, PRECISION]\n",
    "qcols = [MODEL, METHOD] + CORE_METRICS3\n",
    "\n",
    "agg_df_plot_pr_all = df.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD])\n",
    "agg_df_plot_pr_all = agg_df_plot_pr_all.apply(lambda x: x)\n",
    "\n",
    "for x in EXCLUDE:\n",
    "    agg_df_plot_pr_all = agg_df_plot_pr_all.query(f\"method != '{x}'\")    \n",
    "\n",
    "agg_df_small_plot_pr_all = df_small.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD])\n",
    "agg_df_small_plot_pr_all = agg_df_small_plot_pr_all.apply(lambda x: x)\n",
    "\n",
    "for x in EXCLUDE:\n",
    "    agg_df_small_plot_pr_all = agg_df_small_plot_pr_all.query(f\"method != '{x}'\")\n",
    "\n",
    "print(agg_df_plot_pr_all.columns)\n",
    "\n",
    "\n",
    "dfpvals = pd.DataFrame(index=range(9),columns=range(9))\n",
    "print(agg_df_plot_pr_all.loc[:, 'model'].unique())\n",
    "print(agg_df_plot_pr_all.loc[:, 'method'].unique())\n",
    "labels = []\n",
    "labels2 = []\n",
    "count = 0\n",
    "\n",
    "for model in agg_df_plot_pr_all.loc[:, 'model'].unique():  \n",
    "    print(\"count \"+str(count))\n",
    "    for method in agg_df_plot_pr_all.loc[:, 'method'].unique():\n",
    "        #print(\"1 \"+model+\"\\t\"+method)\n",
    "        curmethod = model+\"__\"+method \n",
    "        labels.append(curmethod)\n",
    "        #print((agg_df_plot_pr_all['model'] == model) & (agg_df_plot_pr_all['method'] == method))\n",
    "        vals1 = agg_df_plot_pr_all[(agg_df_plot_pr_all['model'] == model) & (agg_df_plot_pr_all['method'] == method)]\n",
    "        print(vals1.shape)\n",
    "        count2 = 0\n",
    "        first = False\n",
    "        for model2 in agg_df_plot_pr_all.loc[:, 'model'].unique():             \n",
    "            for method2 in agg_df_plot_pr_all.loc[:, 'method'].unique():\n",
    "                print(\"count2 \"+str(count2))                \n",
    "                curmethod2 = model2+\"__\"+method2 \n",
    "                if(not first):\n",
    "                    labels2.append(curmethod2)\n",
    "                if(curmethod != curmethod2):\n",
    "                    print(\"1 \"+model+\"\\t\"+method+\"\\t\"+\"2 \"+model2+\"\\t\"+method2)\n",
    "                    #print((agg_df_plot_pr_all['model'] == model) & (agg_df_plot_pr_all['method'] == method))\n",
    "                    vals2 = agg_df_plot_pr_all[(agg_df_plot_pr_all['model'] == model2) & (agg_df_plot_pr_all['method'] == method2)]\n",
    "                    print(vals2.shape)\n",
    "                    pval= ttest_ind(vals1['precision'], vals2['precision'], alternative='greater', equal_var = False)\n",
    "                    print(pval)\n",
    "                    print(pval[1])\n",
    "                    #print(type(pval[1]))\n",
    "                    dfpvals.iloc[count, count2] = pval[1]\n",
    "                else:\n",
    "                    dfpvals.iloc[count, count2] = 0\n",
    "                count2 = count2 + 1\n",
    "        first = True\n",
    "        count = count + 1\n",
    "\n",
    "dfpvals = dfpvals.apply(pd.to_numeric)\n",
    "dfpvals.index = labels\n",
    "dfpvals.columns = labels\n",
    "\n",
    "dfpvals_correct = dfpvals.copy()\n",
    "dfpvals_correct = multipletests(dfpvals.to_numpy().flatten(), method='fdr_bh')\n",
    "print(type(dfpvals_correct))\n",
    "dfpvals_correct = [*dfpvals_correct[1]]\n",
    "print(type(dfpvals_correct))\n",
    "print(dfpvals_correct)\n",
    "\n",
    "dfpvals_correct_lists = [dfpvals_correct[x:x+9] for x in range(0, len(dfpvals_correct), 9)]#list(it.batched(iterable, n))\n",
    "dfpvals_correct = pd.DataFrame(dfpvals_correct_lists, index =labels, columns =labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b37d763-57ba-4879-b7fd-de52f1b76988",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#for c in dfpvals.select_dtypes(include = [np.number]).columns:\n",
    "#    dfpvals[c] = np.log10(dfpvals[c].values)\n",
    "    \n",
    "ax = plt.axes()\n",
    "#dfpvals_plot = dfpvals\n",
    "#dfpvals_plot[dfpvals_plot < 0.0001] = 1\n",
    "#dfpvals_plot[dfpvals_plot >= 0.0001] = 0\n",
    "\n",
    "#dfpvals[dfpvals < 0.0001] = 0\n",
    "\n",
    "#numeric = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']#\n",
    "#dfpvalslog = dfpvals.copy()\n",
    "#for c in [c for c in dfpvals.columns if dfpvals[c].dtype in numeric]:\n",
    "#    dfpvalslog[c] = np.log10(dfpvals[c])\n",
    "    \n",
    "sns.heatmap(dfpvals_correct, cmap=\"YlGnBu\", annot=True, annot_kws={\"fontsize\":8}, ax = ax)\n",
    "ax.set_title('Precision t-test')\n",
    "\n",
    "plt.savefig(\"methodpairs_precision_correct_ttest.pdf\", format=\"pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0732aae-b4fb-44e5-b3ab-b0c4a450c3f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sns.clustermap(dfpvals_correct, cmap=\"YlGnBu\", annot =True,annot_kws={\"fontsize\":8})\n",
    "\n",
    "ax = g.ax_heatmap\n",
    "\n",
    "rorder = g.dendrogram_row.reordered_ind\n",
    "corder = g.dendrogram_col.reordered_ind\n",
    "\n",
    "#countr = 0\n",
    "#for r in dfpvals.index:\n",
    "#    countc = 0\n",
    "#    for c in dfpvals.columns:\n",
    "#        #print(dfpvals.iloc[countr, countc])\n",
    "#        if(pd.to_numeric(dfpvals.iloc[countr, countc]) < 0.001):\n",
    "#            outf = \"{r:}\\t{c:}\\t\\t\\t{nr:}\\t{nc:}\"\n",
    "#            print(outf.format(r=countr, c=countc, nr=rorder[countr], nc=corder[countc]))\n",
    "#            #len(rorder) - rorder[countr]-1\n",
    "#            ax.add_patch(Rectangle((corder[countc], len(rorder) - rorder[countr]-1), 1, 1, fill=False, edgecolor='orange', lw=2))\n",
    "#        countc=countc+1\n",
    "#    countr=countr+1\n",
    "\n",
    "ax.set_title('Precision t-test')\n",
    "plt.savefig(\"methodpairs_precision_ttest_correct_cluster.pdf\", format=\"pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8541266f-6756-4917-a279-6f5d68672cda",
   "metadata": {},
   "outputs": [],
   "source": [
    "CORE_METRICS3 = [RECALL, PRECISION]\n",
    "qcols = [MODEL, METHOD] + CORE_METRICS3\n",
    "\n",
    "agg_df_plot_pr_all = df.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD])\n",
    "agg_df_plot_pr_all = agg_df_plot_pr_all.apply(lambda x: x)\n",
    "\n",
    "for x in EXCLUDE:\n",
    "    agg_df_plot_pr_all = agg_df_plot_pr_all.query(f\"method != '{x}'\")    \n",
    "\n",
    "agg_df_small_plot_pr_all = df_small.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD])\n",
    "agg_df_small_plot_pr_all = agg_df_small_plot_pr_all.apply(lambda x: x)\n",
    "\n",
    "for x in EXCLUDE:\n",
    "    agg_df_small_plot_pr_all = agg_df_small_plot_pr_all.query(f\"method != '{x}'\")\n",
    "\n",
    "print(agg_df_plot_pr_all.columns)\n",
    "\n",
    "\n",
    "dfpvals = pd.DataFrame(index=range(9),columns=range(9))\n",
    "print(agg_df_plot_pr_all.loc[:, 'model'].unique())\n",
    "print(agg_df_plot_pr_all.loc[:, 'method'].unique())\n",
    "labels = []\n",
    "labels2 = []\n",
    "count = 0\n",
    "\n",
    "for model in agg_df_plot_pr_all.loc[:, 'model'].unique():  \n",
    "    print(\"count \"+str(count))\n",
    "    for method in agg_df_plot_pr_all.loc[:, 'method'].unique():\n",
    "        #print(\"1 \"+model+\"\\t\"+method)\n",
    "        curmethod = model+\"__\"+method \n",
    "        labels.append(curmethod)\n",
    "        #print((agg_df_plot_pr_all['model'] == model) & (agg_df_plot_pr_all['method'] == method))\n",
    "        vals1 = agg_df_plot_pr_all[(agg_df_plot_pr_all['model'] == model) & (agg_df_plot_pr_all['method'] == method)]\n",
    "        print(vals1.shape)\n",
    "        count2 = 0\n",
    "        first = False\n",
    "        for model2 in agg_df_plot_pr_all.loc[:, 'model'].unique():             \n",
    "            for method2 in agg_df_plot_pr_all.loc[:, 'method'].unique():\n",
    "                print(\"count2 \"+str(count2))                \n",
    "                curmethod2 = model2+\"__\"+method2 \n",
    "                if(not first):\n",
    "                    labels2.append(curmethod2)\n",
    "                if(curmethod != curmethod2):\n",
    "                    print(\"1 \"+model+\"\\t\"+method+\"\\t\"+\"2 \"+model2+\"\\t\"+method2)\n",
    "                    #print((agg_df_plot_pr_all['model'] == model) & (agg_df_plot_pr_all['method'] == method))\n",
    "                    vals2 = agg_df_plot_pr_all[(agg_df_plot_pr_all['model'] == model2) & (agg_df_plot_pr_all['method'] == method2)]\n",
    "                    print(vals2.shape)\n",
    "                    pval= ttest_ind(vals1['recall'], vals2['recall'], alternative='greater', equal_var = False)\n",
    "                    print(pval)\n",
    "                    print(pval[1])\n",
    "                    #print(type(pval[1]))\n",
    "                    dfpvals.iloc[count, count2] = pval[1]\n",
    "                else:\n",
    "                    dfpvals.iloc[count, count2] = 0\n",
    "                count2 = count2 + 1\n",
    "        first = True\n",
    "        count = count + 1\n",
    "\n",
    "dfpvals = dfpvals.apply(pd.to_numeric)\n",
    "dfpvals.index = labels\n",
    "dfpvals.columns = labels\n",
    "\n",
    "dfpvals_correct = dfpvals.copy()\n",
    "dfpvals_correct = multipletests(dfpvals.to_numpy().flatten(), method='fdr_bh')\n",
    "print(type(dfpvals_correct))\n",
    "dfpvals_correct = [*dfpvals_correct[1]]\n",
    "print(type(dfpvals_correct))\n",
    "print(dfpvals_correct)\n",
    "\n",
    "dfpvals_correct_lists = [dfpvals_correct[x:x+9] for x in range(0, len(dfpvals_correct), 9)]#list(it.batched(iterable, n))\n",
    "dfpvals_correct = pd.DataFrame(dfpvals_correct_lists, index =labels, columns =labels)\n",
    "\n",
    "dfpvals_correct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ac68500-9c5d-471b-b802-0fbf972f7f09",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ax = plt.axes()\n",
    "#dfpvals_plot = dfpvals\n",
    "#dfpvals_plot[dfpvals_plot < 0.0001] = 1\n",
    "#dfpvals_plot[dfpvals_plot >= 0.0001] = 0\n",
    "\n",
    "#dfpvals[dfpvals < 0.0001] = 0\n",
    "\n",
    "numeric = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']\n",
    "#dfpvalslog = dfpvals.copy()\n",
    "#for c in [c for c in dfpvals.columns if dfpvals[c].dtype in numeric]:\n",
    "#    dfpvalslog[c] = np.log10(dfpvals[c])\n",
    "    \n",
    "sns.heatmap(dfpvals_correct, cmap=\"YlGnBu\", annot=True, annot_kws={\"fontsize\":8}, ax = ax)\n",
    "ax.set_title('Recall t-test')\n",
    "\n",
    "plt.savefig(\"methodpairs_recall_ttest_correct.pdf\", format=\"pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "248308d8-6f2b-4c9d-97c3-da33ee425175",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sns.clustermap(dfpvals_correct, cmap=\"YlGnBu\", annot =True, annot_kws={\"fontsize\":8})\n",
    "\n",
    "ax = g.ax_heatmap\n",
    "\n",
    "rorder = g.dendrogram_row.reordered_ind\n",
    "corder = g.dendrogram_col.reordered_ind\n",
    "\n",
    "#countr = 0\n",
    "#for r in dfpvals.index:\n",
    "#    countc = 0\n",
    "#    for c in dfpvals.columns:\n",
    "#        #print(dfpvals.iloc[countr, countc])\n",
    "#        if(pd.to_numeric(dfpvals.iloc[countr, countc]) < 0.001):\n",
    "#            outf = \"{r:}\\t{c:}\\t\\t\\t{nr:}\\t{nc:}\"\n",
    "#            #print(outf.format(r=countr, c=countc, nr=rorder[countr], nc=corder[countc]))\n",
    "#            #len(rorder) - rorder[countr]-1\n",
    "#            ax.add_patch(Rectangle((corder[countc], len(rorder) - rorder[countr]-1), 1, 1, fill=False, edgecolor='orange', lw=2))\n",
    "#        countc=countc+1\n",
    "#    countr=countr+1\n",
    "\n",
    "ax.set_title('Recall t-test')\n",
    "plt.savefig(\"methodpairs_recall_ttest_correct_cluster.pdf\", format=\"pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78bdc352",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "qcols = [MODEL, METHOD] + CORE_METRICS\n",
    "\n",
    "print(METHOD)\n",
    "#agg_df_orig = df.replace(r\"_\", \" \", regex=True)[qcols].groupby([MODEL, METHOD]).mean(numeric_only=True)\n",
    "#agg_df = pd.DataFrame( columns=agg_df_orig.columns)#NULL#pd.DataFrame() \n",
    "#print(EXCLUDE)\n",
    "#agg_df = agg_df_orig.query(f\"method in @METHOD\")\n",
    "\n",
    "print(qcols)\n",
    "\n",
    "newindex = agg_df_plot.index\n",
    "#len = agg_df.index.to_flat_index().size()\n",
    "#print(len)\n",
    "#for i in range(1,len):\n",
    "#    newindex[i] = agg_df.index[i]\n",
    "\n",
    "#print(agg_df.loc[:, 'model'])\n",
    "\n",
    "#print(agg_df['num GO terms'])\n",
    "\n",
    "#rework for significant GO terms\n",
    "xs = agg_df_plot['has top term']\n",
    "ys = agg_df_plot['proportion significant']\n",
    "xs_er = agg_df_plot_std['has top term']\n",
    "ys_er = agg_df_plot_std['proportion significant']\n",
    "\n",
    "xs_small = agg_df_small_plot['has top term']\n",
    "ys_small = agg_df_small_plot['proportion significant']\n",
    "xs_small_er = agg_df_small_plot_std['has top term']\n",
    "ys_small_er = agg_df_small_plot_std['proportion significant']\n",
    "\n",
    "\n",
    "#print(agg_df.index)\n",
    "\n",
    "print(xs)\n",
    "print(ys)\n",
    "print(np.max(xs))\n",
    "print(np.max(ys))\n",
    "print(np.max(xs_small))\n",
    "print(np.max(ys_small))\n",
    "\n",
    "\n",
    "yells = [\"yellow\"]*5\n",
    "reds = [\"red\"] *3\n",
    "purps = [\"purple\"] *3\n",
    "orans = [\"orange\"]*3\n",
    "\n",
    "colorr = mpl.colors.ColorConverter.to_rgb(\"red\")\n",
    "rgbr = [color_lightness(colorr, scale) for scale in [2]]\n",
    "colorp = mpl.colors.ColorConverter.to_rgb(\"purple\")\n",
    "rgbp = [color_lightness(colorp, scale) for scale in [2]]\n",
    "coloro = mpl.colors.ColorConverter.to_rgb(\"orange\")\n",
    "rgbo = [color_lightness(coloro, scale) for scale in [2]]\n",
    "\n",
    "redss = rgbr *3\n",
    "purpss = rgbp *3\n",
    "oranss = rgbo *3\n",
    "\n",
    "\n",
    "#colors = [*yells, *reds, *purps, *orans]\n",
    "colors = [*reds, *purps, *orans]\n",
    "print(colors)\n",
    "\n",
    "colors_small = [*redss, *purpss, *oranss]\n",
    "print(colors_small)\n",
    "\n",
    "circles= ['o','^','s']*3\n",
    "#triangles= ['^']*3\n",
    "#squares= ['s']*3\n",
    "\n",
    "markers = [*circles]#, *triangles, *squares]\n",
    "\n",
    "SMALL_SIZE = 6\n",
    "plt.rc('font', size=SMALL_SIZE)   \n",
    "plt.figure(figsize=(6,6))\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "#ax.scatter(xs,ys, c=colors, s=50, marker=marker)\n",
    "#ax.scatter(xs_small,xs_small, c=colors, s=25, marker=marker)\n",
    "\n",
    "unique_markers = set(markers)\n",
    "print(unique_markers)\n",
    "print(markers)\n",
    "\n",
    "for um in unique_markers:\n",
    "    #print(um)\n",
    "    #print(markers.index(um))    \n",
    "    #print([index for index, elem in enumerate(markers) if elem == um])\n",
    "    mask = [index for index, elem in enumerate(markers) if elem == um]# markers == um\n",
    "    maskar = np.array(mask)\n",
    "    print(mask)\n",
    "    print(type(mask))\n",
    "    print(maskar)\n",
    "    print(type(maskar))\n",
    "    minv = mask[0]\n",
    "    maxv = mask[2]\n",
    "    print(str(minv)+\"\\t\"+str(maxv))\n",
    "    nowcol = [colors[mask[0]], colors[mask[1]], colors[mask[2]]]\n",
    "    print(nowcol)\n",
    "    \n",
    "    ax.scatter(xs[mask], ys[mask], marker=um, c=nowcol, s=150, alpha=0.5, edgecolors=\"black\",  linewidth=0.5)\n",
    "    ax.scatter(xs_small[mask], ys_small[mask], marker=um, c=nowcol, s=50, alpha=0.5, edgecolors=\"black\",  linewidth=0.5)\n",
    "\n",
    "    ax.errorbar(xs[mask], ys[mask], xerr= xs_er[mask], yerr=ys_er[mask], fmt=\"o\", elinewidth=0.4, capsize=0.4)# ecolor=nowcol,\n",
    "    ax.errorbar(xs_small[mask], ys_small[mask], xerr= xs_small_er[mask], yerr=ys_small_er[mask], fmt=\"o\", elinewidth=0.4, capsize=0.4)\n",
    "    \n",
    "#, label=agg_df.index\n",
    "agg_df_plot_labels = agg_df_plot.index.values\n",
    "for i in range(0, len(xs)):\n",
    "    agg_df_plot_labels[i] = (agg_df_plot_labels[i][0] + \": \" + agg_df_plot_labels[i][1]).replace(\"(\",\"\")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\")\",\"\")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\" - \",\" : \")\n",
    "    agg_df_plot_labels[i] = agg_df_plot_labels[i].replace(\"gpt\",\"GPT\")        \n",
    "                \n",
    "for i in range(0,len(xs)):\n",
    "    ax.annotate(agg_df_plot_labels[i], (xs[i], ys[i]))\n",
    "    ax.annotate(agg_df_plot_labels[i], (xs_small[i], ys_small[i]))  \n",
    "ax.set_xlabel(\"Has top term\")\n",
    "ax.set_ylabel(\"Proportion significant\")\n",
    "\n",
    "#handles, labels = ax.get_legend_handles_labels()\n",
    "#ax.legend(handles, labels, loc='upper left')\n",
    "\n",
    "plt.xlim(-0.2,1)\n",
    "plt.ylim(-0.2,1)\n",
    "plt.gca().set_aspect(\"equal\")\n",
    "\n",
    "plt.savefig(\"top_vs_proportion_std.pdf\", format=\"pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "139178c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "means.query(\"method != 'standard'\").style.highlight_max(axis=0, props='font-weight:bold').format(precision=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b282fac9",
   "metadata": {},
   "source": [
    "## TABLE: evaluation for gene sets < 75"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4eba28b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "agg_table(df_small, CORE_METRICS, EXCLUDE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf6ad663",
   "metadata": {},
   "source": [
    "## Looking at individual gene sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45d76665",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def terms_summary(df, variant=\"v1\", max_rows=9999):\n",
    "    term_dict = {}\n",
    "    n = 0\n",
    "    for _, row in df.iterrows():\n",
    "        if row[PROMPT_VARIANT] and row[PROMPT_VARIANT] != variant:\n",
    "            continue\n",
    "        n += 1\n",
    "        if n > max_rows:\n",
    "            break\n",
    "        model = row['model'] \n",
    "        if \"turbo\" in model:\n",
    "            model = \"turbo\"\n",
    "        elif \"davinci\" in model:\n",
    "            model = \"dav\"\n",
    "        else:\n",
    "            model = \"\"\n",
    "        method = str(row['method']).replace('_', ' ')\n",
    "        if method in [\"closure\", \"rank_based\", \"random\"]:\n",
    "            continue\n",
    "        mm = f\"{model} {method}\"\n",
    "        if method == \"standard\":\n",
    "            nr_term_ids = list(filter_redundant(row[TERM_IDS]))\n",
    "        else:\n",
    "            nr_term_ids = None\n",
    "        for ix, t_id in enumerate(row[TERM_IDS]):\n",
    "            if t_id not in term_dict:\n",
    "                t = {\"id\": t_id, \"label\": go.label(t_id), \"redundant\": False}\n",
    "                term_dict[t_id] = t\n",
    "            else:\n",
    "                t = term_dict[t_id]\n",
    "            t[mm] = ix\n",
    "            if nr_term_ids and t_id not in nr_term_ids:\n",
    "                t[\"redundant\"] = True\n",
    "    objs = list(term_dict.values())\n",
    "    return pd.DataFrame(objs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "383a5109",
   "metadata": {},
   "outputs": [],
   "source": [
    "import oaklib.datamodels.obograph as og\n",
    "from oaklib.utilities.obograph_utils import graph_to_image, default_stylemap_path\n",
    "!mkdir -p output\n",
    "\n",
    "MMAP = {\"standard\": None, \n",
    "        \"gpt-3.5-turbo.no_synopsis\": \"NS\",\n",
    "        \"gpt-3.5-turbo.ontological_synopsis\": \"ONT\",\n",
    "        \"gpt-3.5-turbo.narrative_synopsis\": \"NAR\",\n",
    "       }\n",
    "\n",
    "def viz(geneset, include_std=True, mmap = MMAP, variant=\"v1\"):\n",
    "    methods = mmap.keys()\n",
    "    [gsobj] = [c for c in comps if c.name == geneset]\n",
    "    std = gsobj.payloads[\"standard\"]\n",
    "    t2p = {e.class_id : e.p_value_adjusted for e in std.enrichment_results}\n",
    "    terms = set()\n",
    "    terms.add(\"GO:0008150\")\n",
    "    terms.add(\"GO:0003674\")\n",
    "    terms.add(\"GO:0005575\")\n",
    "    m2t = defaultdict(list)\n",
    "    t2m = defaultdict(set)\n",
    "    seeds = []\n",
    "    for m in methods:\n",
    "        if m == \"standard\":\n",
    "            mv = m\n",
    "        else:\n",
    "            mv = f\"{m}.{variant}\"\n",
    "        p = gsobj.payloads[mv]\n",
    "        if include_std or m != \"standard\":\n",
    "            terms.update(p.term_ids)\n",
    "        for t in p.term_ids:\n",
    "            if t.startswith(\"MONDO:\"):\n",
    "                continue\n",
    "            m2t[m].append(t)\n",
    "            t2m[t].add(m)\n",
    "            if m == \"standard\":\n",
    "                seeds.append(t)\n",
    "    # rels = go.gap_fill_relationships(list(terms), predicates=[IS_A, PART_OF, \"RO:0002211\", \"RO:0002212\", RO:0002213\"])\n",
    "    rels = go.gap_fill_relationships(list(terms), predicates=[IS_A, PART_OF])\n",
    "    g = go.relationships_to_graph(rels)\n",
    "    for n in g.nodes:\n",
    "        if not n.lbl:\n",
    "            n.lbl = go.label(n.id)\n",
    "        if not n.lbl:\n",
    "            n.lbl = n.id\n",
    "        if n.id in t2p:\n",
    "            n.lbl += f\" {t2p[n.id]:.2e}\"\n",
    "        #for m in t2m[n.id]:\n",
    "        #    n.lbl += f\" [<b>{m}</b>]\"\n",
    "    for m in m2t.keys():\n",
    "        if m == \"standard\":\n",
    "            continue\n",
    "        for t in m2t[m]:\n",
    "            n = og.Node(id=f\"{mmap[m]}:{t}\", lbl=mmap[m])\n",
    "            g.nodes.append(n)\n",
    "            g.edges.append(og.Edge(n.id, \"has\", t))\n",
    "    outfile = f\"output/{geneset.replace(' ', '_')}-{include_std}-{variant}.png\"\n",
    "    graph_to_image(g, seeds=seeds, imgfile=outfile, stylemap=\"conf/enr-style.json\")\n",
    "    graph_to_image(g, seeds=seeds, imgfile=outfile, stylemap=\"conf/enr-style.json\")\n",
    "    #return g\n",
    "    \n",
    "viz('peroxisome-0')\n",
    "viz('peroxisome-0', variant=\"v2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c530577b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def geneset_summary(df, geneset):\n",
    "    sdf = terms_summary(df.query(f\"{GENESET} == '{geneset}'\").sort_values(\"similarity\", ascending=False))\n",
    "    [gsobj] = [c for c in comps if c.name == geneset]\n",
    "    std = gsobj.payloads[\"standard\"]\n",
    "    t2p = {e.class_id : e.p_value_adjusted for e in std.enrichment_results}\n",
    "    sdf[\"p_label\"] = sdf.apply(lambda row: str(row.label) + \" \" + str(t2p.get(row.id, \"\")), axis=1)\n",
    "    return sdf\n",
    "\n",
    "geneset_summary(df, 'peroxisome-0')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "195cd45b",
   "metadata": {},
   "source": [
    "### Peroxisome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12f1cc8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "peroxisome = df.query(f\"{GENESET} == 'peroxisome-0'\").sort_values(\"similarity\", ascending=False)\n",
    "peroxisome[[MODEL, METHOD] + eval_summary_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d941a6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# terms_summary(peroxisome).style.highlight_min(axis=1, props='font-weight:bold', numeric_only=True)\n",
    "terms_summary(peroxisome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e635e24",
   "metadata": {},
   "outputs": [],
   "source": [
    "terms_summary(peroxisome, \"v2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "190f06b0",
   "metadata": {},
   "source": [
    "## Sensory Ataxia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1a7517a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ataxia = df.query(f\"{GENESET} == 'sensory ataxia-0'\").sort_values(\"similarity\", ascending=False)\n",
    "ataxia[[MODEL, METHOD] + eval_summary_cols]                                                  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d62ed4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_colwidth', None)\n",
    "pd.set_option('display.max_rows', None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "848fbb12",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ataxia[[MODEL, METHOD, PROMPT_VARIANT, GO_TERM_IDS, NOVEL_LABELS]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56431e67",
   "metadata": {},
   "outputs": [],
   "source": [
    "viz('sensory ataxia-0')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8973c4ce",
   "metadata": {},
   "source": [
    "![img](output/sensory_ataxia-0-True-v1.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adc82467",
   "metadata": {},
   "outputs": [],
   "source": [
    "viz('sensory ataxia-0', variant=\"v2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e599f14b",
   "metadata": {},
   "source": [
    "![img](output/sensory_ataxia-0-True-v2.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb7c8e15",
   "metadata": {},
   "outputs": [],
   "source": [
    "terms_summary(ataxia)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57bf6fdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def retrieve_payload(geneset, method):\n",
    "    for comp in comps:\n",
    "        if comp.name == geneset:\n",
    "            return comp.payloads[method]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48a0a117",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(retrieve_payload(\"sensory ataxia-0\", \"gpt-3.5-turbo.ontological_synopsis.v1\").response_text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68810d3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(retrieve_payload(\"sensory ataxia-0\", \"gpt-3.5-turbo.narrative_synopsis.v1\").response_text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed7e6d7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(retrieve_payload(\"sensory ataxia-0\", \"gpt-3.5-turbo.no_synopsis.v1\").response_text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8ca6578",
   "metadata": {},
   "source": [
    "## T cell proliferation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15940214",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "tcp = df.query(f\"{GENESET} == 'T cell proliferation-0'\").sort_values(\"similarity\", ascending=False)\n",
    "tcp[[MODEL, METHOD] + eval_summary_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "418db0bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "viz('T cell proliferation-0')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8ab594f",
   "metadata": {},
   "source": [
    "![img](output/T_cell_proliferation-0-True-v1.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e200c280",
   "metadata": {},
   "outputs": [],
   "source": [
    "terms_summary(tcp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71312d76",
   "metadata": {},
   "source": [
    "## Endocytosis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24b60582",
   "metadata": {},
   "outputs": [],
   "source": [
    "endocytosis = df.query(f\"{GENESET} == 'endocytosis-0'\").sort_values(\"similarity\", ascending=False)\n",
    "terms_summary(endocytosis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "051a7dfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(retrieve_payload(\"endocytosis-0\", \"gpt-3.5-turbo.narrative_synopsis.v1\").prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a4b48ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(retrieve_payload(\"endocytosis-0\", \"gpt-3.5-turbo.narrative_synopsis.v1\").response_text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41948b07",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(retrieve_payload(\"endocytosis-0\", \"gpt-3.5-turbo.ontological_synopsis.v1\").prompt)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9405a5c6",
   "metadata": {},
   "source": [
    "## Hydrolysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e014a6dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "hydrolysis = df.query(f\"{GENESET} == 'hydrolase activity, hydrolyzing O-glycosyl compounds-0'\").sort_values(\"similarity\", ascending=False)\n",
    "terms_summary(hydrolysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bf11492",
   "metadata": {},
   "outputs": [],
   "source": [
    "viz('hydrolase activity, hydrolyzing O-glycosyl compounds-0')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5bbe239",
   "metadata": {},
   "source": [
    "![img](output/hydrolase_activity,_hydrolyzing_O-glycosyl_compounds-0-True-v1.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e40e3678",
   "metadata": {},
   "source": [
    "## Variability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daac5fd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pv_pivot = df.pivot_table(index=[MODEL, METHOD, GENESET], columns=PROMPT_VARIANT, values=PROPORTION_SIGNIFICANT)\n",
    "# Calculate differences between run \"1\" and run \"2\"\n",
    "pv_pivot['diff'] = pv_pivot[\"v1\"] - pv_pivot[\"v2\"]\n",
    "pv_pivot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97a7b1fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Now you can perform statistics on the 'diff' column\n",
    "df_diff_stats = pv_pivot['diff'].agg(['mean', 'std', 'var', 'min', 'max'])\n",
    "df_diff_stats['range'] = df_diff_stats['max'] - df_diff_stats['min']\n",
    "df_diff_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36d1586a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def jaccard_similarity(list1, list2):\n",
    "    intersection = len(list(set(list1).intersection(list2)))\n",
    "    union = (len(list1) + len(list2)) - intersection\n",
    "    return float(intersection) / union\n",
    "\n",
    "# pivot your dataframe as before\n",
    "df_pivot = df.pivot_table(index=[MODEL, METHOD, GENESET], columns=PROMPT_VARIANT, values=TERM_IDS, aggfunc=list)\n",
    "df_pivot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2583c45",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# calculate Jaccard index for each group\n",
    "df_pivot['jaccard_index'] = df_pivot.apply(lambda row: jaccard_similarity(row[\"v1\"][0], row[\"v2\"][0]), axis=1)\n",
    "\n",
    "# reset index to get it back in the form of a DataFrame\n",
    "result = df_pivot.reset_index()\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e27f0eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "result['jaccard_index'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eff5ac1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "result.groupby([MODEL])['jaccard_index'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "655a81d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "result.groupby([MODEL, METHOD])['jaccard_index'].describe()[['count', 'mean', 'std', 'min', 'max']].style.highlight_max(axis=0, props='font-weight:bold').format(precision=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60911ad3",
   "metadata": {},
   "outputs": [],
   "source": [
    "summaries = [s for s in list(df[SUMMARY]) if s]\n",
    "len(summaries)\n",
    "import random\n",
    "def random_summary():\n",
    "    return summaries[int(random.random() * len(summaries))]\n",
    "\n",
    "random_summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7db7ed82",
   "metadata": {},
   "outputs": [],
   "source": [
    "from ontogpt.clients import OpenAIClient\n",
    "\n",
    "simclient = OpenAIClient(model=\"text-embedding-ada-002\")\n",
    "\n",
    "def text_similarity(text1, text2):\n",
    "    return simclient.similarity(text1, text2)\n",
    "\n",
    "text_similarity(\"nucleus of cell\", \"nuclear membrane\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b702b36",
   "metadata": {},
   "outputs": [],
   "source": [
    "text_similarity(random_summary(), random_summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0fbac9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "rs = random_summary()\n",
    "text_similarity(rs, rs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fba61927",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pivot = df.pivot_table(index=[MODEL, METHOD, GENESET], columns=PROMPT_VARIANT, values=SUMMARY, aggfunc=list)\n",
    "df_pivot\n",
    "\n",
    "df_pivot['sim'] = df_pivot.apply(lambda row: text_similarity(row[\"v1\"][0], row[\"v2\"][0]), axis=1)\n",
    "df_pivot['length_diff'] = df_pivot.apply(lambda row: abs(len(row[\"v1\"][0])-len(row[\"v2\"][0])), axis=1)\n",
    "\n",
    "\n",
    "# reset index to get it back in the form of a DataFrame\n",
    "result = df_pivot.reset_index()\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "281816a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnd = result.query(\"model=='text-davinci-003' and method=='no_synopsis'\")\n",
    "rnd[MODEL]=\"\"\n",
    "rnd[METHOD]=\"RANDOM\"\n",
    "rnd[\"sim\"] = rnd.apply(lambda row: text_similarity(random_summary(), random_summary()), axis=1)\n",
    "rnd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8881793a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_summary = pd.concat([result, rnd]).groupby([MODEL, METHOD])['sim'].describe()[['count', 'mean', 'std', 'min', 'max']]\n",
    "sim_summary.style.highlight_max(axis=0, props='font-weight:bold').format(precision=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a71dfb59",
   "metadata": {},
   "outputs": [],
   "source": [
    "stats_df = result.groupby(METHOD).agg({'sim': [np.mean, np.std]}).reset_index()\n",
    "stats_df.columns = ['category', 'mean', 'stddev']\n",
    "\n",
    "# Set the plot style\n",
    "sns.set(style=\"whitegrid\")\n",
    "\n",
    "# Create the bar plot with error bars\n",
    "plt.figure(figsize=(10, 6))\n",
    "bar_plot = sns.barplot(x='category', y='mean', data=stats_df, yerr=stats_df['stddev'], capsize=.1)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel(\"Category\")\n",
    "plt.ylabel(\"Mean Score\")\n",
    "plt.title(\"Mean Score with Standard Deviation for Each Category\")\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "529b7e57",
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_summary = result.groupby([MODEL, METHOD])['length_diff'].describe()[['count', 'mean', 'std', 'min', 'max']]\n",
    "diff_summary.style.highlight_max(axis=0, props='font-weight:bold').format(precision=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a739315c",
   "metadata": {},
   "source": [
    "## Unparsed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "301b7b9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def unparsed(df, model):\n",
    "    novel_term_map = defaultdict(dict)\n",
    "    for _, row in df.iterrows():\n",
    "        gs = row[GENESET]\n",
    "        #if not gs.endswith(\"-0\"):\n",
    "        #    continue\n",
    "        gs = gs.replace(\"-0\", \"\")\n",
    "        for lbl in row[NOVEL_LABELS]:\n",
    "            if row[MODEL] != model:\n",
    "                continue\n",
    "            m = row[METHOD]\n",
    "            novel_term_map[lbl][\"NAME\"] = lbl\n",
    "            if gs not in novel_term_map[lbl]:\n",
    "                novel_term_map[lbl][m] = []\n",
    "            novel_term_map[lbl][m].append(gs)\n",
    "    novel_df = pd.DataFrame(novel_term_map.values())\n",
    "    return novel_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa7f4d70",
   "metadata": {},
   "source": [
    "## Potential Hallucinations\n",
    "\n",
    "Summarize all GO terms that are in all summaries that are not in the closure of annotated terms for any of the genes in the gene set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fe25ede",
   "metadata": {},
   "outputs": [],
   "source": [
    "def hallucinatons(df, model):\n",
    "    novel_term_map = defaultdict(dict)\n",
    "    for _, row in df.iterrows():\n",
    "        gs = row[GENESET]\n",
    "        #if not gs.endswith(\"-0\"):\n",
    "        #    continue\n",
    "        gs = gs.replace(\"-0\", \"\")\n",
    "        for lbl in row[NOVEL_LABELS]:\n",
    "            if row[MODEL] != model:\n",
    "                continue\n",
    "            m = row[METHOD]\n",
    "            novel_term_map[lbl][\"NAME\"] = lbl\n",
    "            if gs not in novel_term_map[lbl]:\n",
    "                novel_term_map[lbl][m] = []\n",
    "            novel_term_map[lbl][m].append(gs)\n",
    "    novel_df = pd.DataFrame(novel_term_map.values())\n",
    "    return novel_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb90b5ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "novel_df_turbo = hallucinatons(df, TURBO).reset_index(drop=True)\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "pd.set_option('display.max_rows', None)\n",
    "novel_df_turbo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da5e097f",
   "metadata": {},
   "outputs": [],
   "source": [
    "hallucinatons(df, DAVINCI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c811184e",
   "metadata": {},
   "outputs": [],
   "source": [
    "hallucinatons(df, GPT4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8605a613",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tiktoken\n",
    "enc = tiktoken.encoding_for_model(\"gpt-4\")\n",
    "enc.encode(\"negative regulation\")\n",
    "\n",
    "def tok_similarity(t1, t2):\n",
    "    if t1 is None and t2 is None:\n",
    "        return 1\n",
    "    if t1 is None or t2 is None:\n",
    "        return 0\n",
    "\n",
    "    toks1 = set(enc.encode(t1))\n",
    "    toks2 = set(enc.encode(t2))\n",
    "    return len(toks1.intersection(toks2)) / len(toks1.union(toks2))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56ce46bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "closure_by_gene_set = {}\n",
    "for comp in comps:\n",
    "    closure = comp.payloads[\"closure\"].term_strings\n",
    "    closure_by_gene_set[comp.name] = closure\n",
    "\n",
    "def hallucinatons_with_best_match(df, model):\n",
    "    novel_term_map = defaultdict(dict)\n",
    "    for _, row in df.iterrows():\n",
    "        gs = row[GENESET]\n",
    "        # print(gs)\n",
    "        closure = closure_by_gene_set[gs]\n",
    "        #if not gs.endswith(\"-0\"):\n",
    "        #    continue\n",
    "        gs = gs.replace(\"-0\", \"\")\n",
    "        for lbl in row[NOVEL_LABELS]:\n",
    "            if row[MODEL] != model:\n",
    "                continue\n",
    "            m = row[METHOD]\n",
    "            novel_term_map[lbl][\"NAME\"] = lbl\n",
    "            if gs not in novel_term_map[lbl]:\n",
    "                novel_term_map[lbl][m] = []\n",
    "            # sims = [(ct, text_similarity(ct, lbl)) for ct in closure]\n",
    "            sims = [(ct, tok_similarity(ct, lbl)) for ct in closure]\n",
    "            sims = sorted(sims, key=lambda x: -x[1])\n",
    "            best = sims[0][0]\n",
    "            novel_term_map[lbl][m].append((gs, best))\n",
    "    novel_df = pd.DataFrame(novel_term_map.values())\n",
    "    return novel_df\n",
    "\n",
    "novel_df_turbo = hallucinatons_with_best_match(df, TURBO).reset_index(drop=True)\n",
    "novel_df_turbo"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86905ba9",
   "metadata": {},
   "source": [
    "## New Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ee36fa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "endocytosis = df.query(f\"{GENESET} == 'endocytosis-0'\").sort_values(\"similarity\", ascending=False)\n",
    "terms_summary(endocytosis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4407a14",
   "metadata": {},
   "outputs": [],
   "source": [
    "viz('endocytosis-0')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddb91a56",
   "metadata": {},
   "source": [
    "![img](output/endocytosis-0-True-v1.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "019bdbd6",
   "metadata": {},
   "source": [
    "### New Annotations\n",
    "\n",
    "in 2022-03-24, GO `molecular sequesting` only had 6 genes annotated; this increased to 30 in 2023.\n",
    "If these are past the LLM training date then we would not expect these to influence results. Additionally,\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdf9d3bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequestering = df.query(f\"{GENESET} == 'molecular sequestering-0'\").sort_values(\"similarity\", ascending=False)\n",
    "sequestering[eval_summary_cols]                                                           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80a297a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "terms_summary(sequestering)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "362de09e",
   "metadata": {},
   "outputs": [],
   "source": [
    "viz('molecular sequestering-0')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbd1074f",
   "metadata": {},
   "source": [
    "![img](output/molecular_sequestering-0-True-v1.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fee93f7d",
   "metadata": {},
   "source": [
    "## IGRB\n",
    "\n",
    "This gene set contains genes previously annotated to IGRB, many of which have been since removed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cda2a2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "igrb = df.query(f\"{GENESET} == 'ig-receptor-binding-2022-0'\").sort_values(\"similarity\", ascending=False)\n",
    "terms_summary(igrb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a374a6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "viz('ig-receptor-binding-2022-0')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c2f02a8",
   "metadata": {},
   "source": [
    "![img](output/ig-receptor-binding-2022-0-True-v1.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93e71cf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "viz('HALLMARK_GLYCOLYSIS-0')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6955dddd",
   "metadata": {},
   "source": [
    "![img](output/HALLMARK_GLYCOLYSIS-0-True-v1.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca525207",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b97d18bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "viz('HALLMARK_KRAS_SIGNALING_UP-0')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a89fb38e",
   "metadata": {},
   "source": [
    "![img](output/HALLMARK_KRAS_SIGNALING_UP-0-True-v1.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "234aa447",
   "metadata": {},
   "source": [
    "## Summaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07cbd616",
   "metadata": {},
   "outputs": [],
   "source": [
    "objs = []\n",
    "for c in comps:\n",
    "    for m, payload in c.payloads.items():\n",
    "        if payload.summary:\n",
    "            objs.append({\"model\": c.model, \"geneset\": c.name, \"method\": m, \"summary\": payload.summary})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7bab9b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sdf = pd.DataFrame(objs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e81f606",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_colwidth', None)\n",
    "pd.set_option('display.max_rows', None)\n",
    "sdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4ce2cbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "sdf.query(\"geneset == 'sensory ataxia-0'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9543bfca",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25a14c32",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c95b3902",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70195fef",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
